{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": "true"
   },
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Introduction\"><span class=\"toc-item-num\">1 - </span>Introduction</a></div><div class=\"lev1\"><a href=\"#Folder-initialization-&amp;-imports\"><span class=\"toc-item-num\">2 - </span>Folder initialization &amp; imports</a></div><div class=\"lev1\"><a href=\"#Mapping-to-experimental-structures\"><span class=\"toc-item-num\">3 - </span>Mapping to experimental structures</a></div><div class=\"lev2\"><a href=\"#Loading-the-GEM\"><span class=\"toc-item-num\">3.1 - </span>Loading the GEM</a></div><div class=\"lev2\"><a href=\"#Dataframe-creation\"><span class=\"toc-item-num\">3.2 - </span>Dataframe creation</a></div><div class=\"lev3\"><a href=\"#Mapping-gene-IDs-to-UniProt-IDs\"><span class=\"toc-item-num\">3.2.1 - </span>Mapping gene IDs to UniProt IDs</a></div><div class=\"lev4\"><a href=\"#Manually-mapping-ones-that-didn't-map\"><span class=\"toc-item-num\">3.2.1.1 - </span>Manually mapping ones that didn't map</a></div><div class=\"lev3\"><a href=\"#Mapping-UniProt-IDs-to-PDB-IDs\"><span class=\"toc-item-num\">3.2.2 - </span>Mapping UniProt IDs to PDB IDs</a></div><div class=\"lev4\"><a href=\"#Alternate-mapping:-BLASTing-the-PDB\"><span class=\"toc-item-num\">3.2.2.1 - </span>Alternate mapping: BLASTing the PDB</a></div><div class=\"lev3\"><a href=\"#Creating-and-saving-the-mapping-Dataframe\"><span class=\"toc-item-num\">3.2.3 - </span>Creating and saving the mapping Dataframe</a></div><div class=\"lev3\"><a href=\"#Obtaining-sequence-metadata\"><span class=\"toc-item-num\">3.2.4 - </span>Obtaining sequence metadata</a></div><div class=\"lev3\"><a href=\"#Obtaining-structure-metadata\"><span class=\"toc-item-num\">3.2.5 - </span>Obtaining structure metadata</a></div><div class=\"lev3\"><a href=\"#Final-integrated-DataFrame\"><span class=\"toc-item-num\">3.2.6 - </span>Final integrated DataFrame</a></div><div class=\"lev1\"><a href=\"#Mapping-to-homology-models\"><span class=\"toc-item-num\">4 - </span>Mapping to homology models</a></div><div class=\"lev2\"><a href=\"#Loading-existing-gene--&gt;-homology-model-dataframe\"><span class=\"toc-item-num\">4.1 - </span>Loading existing gene -&gt; homology model dataframe</a></div><div class=\"lev2\"><a href=\"#Merging-with-the-created-GEM-PRO\"><span class=\"toc-item-num\">4.2 - </span>Merging with the created GEM-PRO</a></div><div class=\"lev1\"><a href=\"#Ranking-experimental-structures\"><span class=\"toc-item-num\">5 - </span>Ranking experimental structures</a></div><div class=\"lev2\"><a href=\"#Scoring-sequence-identity\"><span class=\"toc-item-num\">5.1 - </span>Scoring sequence identity</a></div><div class=\"lev3\"><a href=\"#Writing-FASTA-files-for-the-original-sequence\"><span class=\"toc-item-num\">5.1.1 - </span>Writing FASTA files for the original sequence</a></div><div class=\"lev3\"><a href=\"#Writing-FASTA-files-for-the-structures\"><span class=\"toc-item-num\">5.1.2 - </span>Writing FASTA files for the structures</a></div><div class=\"lev3\"><a href=\"#Aligning-the-structures-to-the-sequence\"><span class=\"toc-item-num\">5.1.3 - </span>Aligning the structures to the sequence</a></div><div class=\"lev3\"><a href=\"#Scoring\"><span class=\"toc-item-num\">5.1.4 - </span>Scoring</a></div><div class=\"lev2\"><a href=\"#Scoring-resolution\"><span class=\"toc-item-num\">5.2 - </span>Scoring resolution</a></div><div class=\"lev2\"><a href=\"#Merging-scores-&amp;-ranking\"><span class=\"toc-item-num\">5.3 - </span>Merging scores &amp; ranking</a></div><div class=\"lev1\"><a href=\"#Saving-the-final-GEM-PRO-dataframe\"><span class=\"toc-item-num\">6 - </span>Saving the final GEM-PRO dataframe</a></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "<span style=\"background-color:#66FF99\">**This IPython notebook will guide you through the creation of a GEM-PRO, for the *E. coli* model iJO1366.**</span>\n",
    "\n",
    "There are three main stages to this workflow.\n",
    "\n",
    "1. Mapping gene IDs to experimentally solved protein structures, available in the Protein Data Bank (PDB)\n",
    "2. Creating new or mapping to existing homology models for genes that do not have an experimentally solved structure.\n",
    "3. Ranking available experimentally solved structures.\n",
    "\n",
    "Following notebooks will outline how we can use this GEM-PRO for strain specific comparisons through 55 *E. coli* strains."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Folder initialization & imports\n",
    "\n",
    "<span style=\"background-color:#66FF99\">**Before we begin, we are going to create the folder structure for our files. **</span>\n",
    "\n",
    "Also, there are a couple of packages required (already installed on the 212 machine). If you are planning to use your own machine, let me know!\n",
    "\n",
    "- Bioservices for querying various REST services\n",
    "        pip install bioservices\n",
    "\n",
    "- libsbml for reading SBML models\n",
    "        pip install python-libsbml\n",
    "        \n",
    "- Biopython and ProDy for various structural methods\n",
    "        pip install prody\n",
    "        pip install biopython (https://github.com/nmih/biopython contains an altered version for scripts in stage 3)\n",
    "        \n",
    "- Other\n",
    "        pip install xmltodict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:#66FF99\">**SSBIO**</span>\n",
    "\n",
    "This notebook contains a variety of imports and functions that are used throughout the workflow. Functions that are used throughout the workflow or those that have a general use are stored in *ssbio*. More specific functions, such as the manual adder below, are isolated in their respective notebooks. For detailed information on the contents of *ssbio*, simply open the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "@> ProDy is configured: verbosity='none'\n",
      "INFO:.prody:ProDy is configured: verbosity='none'\n"
     ]
    }
   ],
   "source": [
    "%run 'ssbio_01.ipynb'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "GEM_FILE_TEMP = './iJO1366.xml'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "GEM_NAME = 'iJO1366'\n",
    "GEM_GENE_SOURCE = 'ENSEMBLGENOME_ID'\n",
    "UNIPROT_REVIEWED_ONLY = True\n",
    "ROOT_DIR = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "MODEL_DIR = os.path.join(ROOT_DIR, GEM_NAME)\n",
    "\n",
    "# data_frames - directory where all data frames will be stored (all stages)\n",
    "DATA_FRAMES = os.path.join(MODEL_DIR, 'data_frames')\n",
    "\n",
    "# model_files - directory where original GEMs and GEM-related files are stored\n",
    "MODEL_FILES = os.path.join(MODEL_DIR, 'model_files')\n",
    "\n",
    "# sequence_files - directory where all sequence related files are stored\n",
    "SEQ_FILES = os.path.join(MODEL_DIR, 'sequence_files')\n",
    "SEQ_UNIPROT_FILES = os.path.join(SEQ_FILES, 'uniprot_sequences')\n",
    "SEQ_PDB_FILES = os.path.join(SEQ_FILES, 'pdb_sequences')\n",
    "SEQ_ALIGN_FILES = os.path.join(SEQ_FILES, 'alignment')\n",
    "\n",
    "# structure_files - directory where structure related files will be downloaded/are located\n",
    "STRUCT_FILES = os.path.join(MODEL_DIR, 'structure_files')\n",
    "STRUCT_EXP_FILES = os.path.join(STRUCT_FILES, 'experimental')\n",
    "STRUCT_HOMOLOGY_FILES = os.path.join(STRUCT_FILES, 'homology_models')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for directory in [MODEL_DIR, DATA_FRAMES, MODEL_FILES, \n",
    "                  SEQ_FILES, SEQ_UNIPROT_FILES, SEQ_PDB_FILES, SEQ_ALIGN_FILES,\n",
    "                  STRUCT_FILES, STRUCT_EXP_FILES, STRUCT_HOMOLOGY_FILES]:\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "\n",
    "shutil.copy(GEM_FILE_TEMP, MODEL_FILES)\n",
    "GEM_FILE = os.path.join(MODEL_FILES, os.path.basename(GEM_FILE_TEMP))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Mapping to experimental structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:#66FF99\">**The first step of this workflow is to utilize ID mapping services to map gene IDs to:**</span>\n",
    "\n",
    "1. The UniProt sequence database\n",
    "2. Available protein structures in the PDB\n",
    "\n",
    "<span style=\"background-color:orange\">**NOTE: students can skip parts of this for their \"base\" models, we will provide the mapping dataframes of gene to sequence to available structures. Parts that can be skipped will be highlighted in orange.**</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading the GEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from cobra import Model, Reaction, Metabolite\n",
    "from cobra.io.sbml import create_cobra_model_from_sbml_file\n",
    "\n",
    "model = create_cobra_model_from_sbml_file(GEM_FILE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Dataframe creation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:#66FF99\">**All data will be stored in a Pandas Dataframe, where a column starting with:**</span>\n",
    "* m_ is information that comes directly from the GEM\n",
    "* u_ is information that comes from UniProt\n",
    "* p_ is information that comes from the PDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping gene IDs to UniProt IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# # mapping gene IDs to uniprot IDs\n",
    "\n",
    "gene_ids = []\n",
    "for gene in model.genes:\n",
    "    gene_ids.append(gene.id)\n",
    "genes_to_uniprots = bsup.mapping(fr=GEM_GENE_SOURCE, to='ACC', query=gene_ids)\n",
    "genes_to_uniprots_missing = list(set(gene_ids).difference(genes_to_uniprots.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['b2092']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_to_uniprots_missing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manually mapping ones that didn't map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter UniProt ID for b2092 ('skip' to ignore): P69831\n"
     ]
    }
   ],
   "source": [
    "def manual_adder(no_mapping, mapping_dict):\n",
    "    final_missing = []\n",
    "    for m_gene in no_mapping:\n",
    "        manual_uniprot = raw_input('Enter UniProt ID for %s (\\'skip\\' to ignore): ' % m_gene)\n",
    "        if manual_uniprot.lower() == 'skip':\n",
    "            final_missing.append(m_gene)\n",
    "            continue\n",
    "        while not uniprot_valid_id(manual_uniprot):\n",
    "            warnings.warn('WARNING: invalid UniProt ID, enter again:')\n",
    "            manual_uniprot = raw_input('Enter UniProt ID for %s (\\'skip\\' to ignore): ' % m_gene)\n",
    "        mapping_dict[m_gene] = [manual_uniprot]\n",
    "        \n",
    "    return mapping_dict, final_missing\n",
    "\n",
    "genes_to_uniprots, final_missing = manual_adder(genes_to_uniprots_missing, genes_to_uniprots)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we can't map these\n",
    "final_missing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping UniProt IDs to PDB IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "uniprots = flatlist_dropdup(genes_to_uniprots.values())\n",
    "uniprots_to_pdbs = bioservices_uniprot_mapper('ACC', 'PDB_ID', uniprots)\n",
    "uniprots_status = uniprot_reviewed_checker(uniprots)\n",
    "unreviewed_uniprots = dict((k, v) for k, v in uniprots_status.iteritems() if v != 'reviewed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# gathering obsolete and theoretical PDBs\n",
    "all_pdbs = flatlist_dropdup(uniprots_to_pdbs.values())\n",
    "pdbs_status = pdb_current_checker(all_pdbs)\n",
    "discard_pdbs = dict((k, v) for k, v in pdbs_status.iteritems() if v == 'theoretical' or v == 'obsolete')\n",
    "obsolete_pdbs_to_new = dict((k, v) for k, v in pdbs_status.iteritems() if isinstance(v,list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating and saving the mapping Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# filtering and creating a mapping dataframe\n",
    "DF_MAP_GENE_UNIPROT_PDB = pd.DataFrame(columns=['m_gene', 'u_uniprot_acc', 'u_reviewed', 'u_pdb'])\n",
    "\n",
    "for gene,uniprots in genes_to_uniprots.iteritems():\n",
    "    \n",
    "    # filtering out unreviewed UniProt entries (if specified)\n",
    "    if UNIPROT_REVIEWED_ONLY:\n",
    "        uniprots = list(set(uniprots).difference(unreviewed_uniprots.keys()))\n",
    "    \n",
    "    if not uniprots:\n",
    "        appender = {}\n",
    "        appender['m_gene'] = gene\n",
    "        DF_MAP_GENE_UNIPROT_PDB = DF_MAP_GENE_UNIPROT_PDB.append(appender, ignore_index=True)\n",
    "    \n",
    "    for uniprot in uniprots:\n",
    "        if uniprot in unreviewed_uniprots:\n",
    "            u_reviewed = False\n",
    "        else:\n",
    "            u_reviewed = True\n",
    "        \n",
    "        if uniprot in uniprots_to_pdbs:\n",
    "            pdbs = uniprots_to_pdbs[uniprot]\n",
    "            \n",
    "            # next, filtering out theoretical PDB models and replacing obsolete IDs with new ones\n",
    "            pdbs_filter = list(set(pdbs).difference(discard_pdbs.keys()))\n",
    "            obsoletes = list(set(pdbs).intersection(obsolete_pdbs_to_new.keys()))\n",
    "            for obsolete in obsoletes:\n",
    "                replacer = obsolete_pdbs_to_new[obsolete]\n",
    "                pdbs_filter.remove(obsolete)\n",
    "                pdbs_filter.extend(replacer)\n",
    "            \n",
    "            pdbs_filter = list(set(pdbs_filter))\n",
    "            \n",
    "            if not pdbs_filter:\n",
    "                pdbs_filter.append(np.nan)\n",
    "            \n",
    "            for pdb in pdbs_filter:\n",
    "                appender = {}\n",
    "                appender['m_gene'] = gene\n",
    "                appender['u_uniprot_acc'] = uniprot\n",
    "                appender['u_reviewed'] = u_reviewed\n",
    "                appender['u_pdb'] = pdb\n",
    "                DF_MAP_GENE_UNIPROT_PDB = DF_MAP_GENE_UNIPROT_PDB.append(appender, ignore_index=True)\n",
    "                \n",
    "        else:\n",
    "            appender = {}\n",
    "            appender['m_gene'] = gene\n",
    "            appender['u_uniprot_acc'] = uniprot\n",
    "            appender['u_reviewed'] = u_reviewed\n",
    "            DF_MAP_GENE_UNIPROT_PDB = DF_MAP_GENE_UNIPROT_PDB.append(appender, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_gene</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>u_pdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b3821</td>\n",
       "      <td>P0A921</td>\n",
       "      <td>True</td>\n",
       "      <td>1ILD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b3821</td>\n",
       "      <td>P0A921</td>\n",
       "      <td>True</td>\n",
       "      <td>1FW2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b3821</td>\n",
       "      <td>P0A921</td>\n",
       "      <td>True</td>\n",
       "      <td>1FW3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b3821</td>\n",
       "      <td>P0A921</td>\n",
       "      <td>True</td>\n",
       "      <td>1QD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b3821</td>\n",
       "      <td>P0A921</td>\n",
       "      <td>True</td>\n",
       "      <td>1QD5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  m_gene u_uniprot_acc u_reviewed u_pdb\n",
       "0  b3821        P0A921       True  1ILD\n",
       "1  b3821        P0A921       True  1FW2\n",
       "2  b3821        P0A921       True  1FW3\n",
       "3  b3821        P0A921       True  1QD6\n",
       "4  b3821        P0A921       True  1QD5"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_MAP_GENE_UNIPROT_PDB.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_MAP_GENE_UNIPROT_PDB.to_csv(os.path.join(DATA_FRAMES, 'DF_MAP_GENE_UNIPROT_PDB.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_MAP_GENE_UNIPROT_PDB = pd.read_csv(os.path.join(DATA_FRAMES, 'DF_MAP_GENE_UNIPROT_PDB.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:RED\">**ABOVE DATAFRAME -- to be provided**</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Obtaining sequence metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "uniprot_accs = list(DF_MAP_GENE_UNIPROT_PDB[DF_MAP_GENE_UNIPROT_PDB.u_uniprot_acc.notnull()].u_uniprot_acc.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***PROGRESS: P0C8K0 - PARSED 1367/1367 UNIPROT IDS***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "uniprot_metadata_all = uniprot_metadata(uniprot_accs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_METADATA_UNIPROT = pd.DataFrame.from_dict(uniprot_metadata_all, orient='index')\n",
    "cols = ['u_uniprot_acc', 'u_gene_name', 'u_ec_number', 'u_description', 'u_seq', 'u_seq_len',\n",
    "        'u_pfam', 'u_kegg_id', 'u_refseq', 'u_go', 'u_reviewed']\n",
    "DF_METADATA_UNIPROT = DF_METADATA_UNIPROT[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_METADATA_UNIPROT.to_csv(os.path.join(DATA_FRAMES, 'DF_METADATA_UNIPROT.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_gene_name</th>\n",
       "      <th>u_ec_number</th>\n",
       "      <th>u_description</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_pfam</th>\n",
       "      <th>u_kegg_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_go</th>\n",
       "      <th>u_reviewed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O32583</th>\n",
       "      <td>O32583</td>\n",
       "      <td>thiS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Thiamine biosynthesis protein ThiS', 'Sulfur...</td>\n",
       "      <td>MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...</td>\n",
       "      <td>66</td>\n",
       "      <td>['PF02597']</td>\n",
       "      <td>['ecj:JW3955', 'eco:b4407']</td>\n",
       "      <td>['WP_001166226.1', 'NZ_LN832404.1', 'YP_026279...</td>\n",
       "      <td>['GO:0000166; F:nucleotide binding; IEA:UniPro...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00350</th>\n",
       "      <td>P00350</td>\n",
       "      <td>gnd</td>\n",
       "      <td>['1.1.1.44']</td>\n",
       "      <td>['6-phosphogluconate dehydrogenase, decarboxyl...</td>\n",
       "      <td>MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...</td>\n",
       "      <td>468</td>\n",
       "      <td>['PF00393', 'PF03446']</td>\n",
       "      <td>['ecj:JW2011', 'eco:b2029']</td>\n",
       "      <td>['NP_416533.1', 'NC_000913.3', 'WP_000043484.1...</td>\n",
       "      <td>['GO:0005829; C:cytosol; IDA:EcoCyc.', 'GO:004...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00363</th>\n",
       "      <td>P00363</td>\n",
       "      <td>frdA</td>\n",
       "      <td>['1.3.5.4']</td>\n",
       "      <td>['Fumarate reductase flavoprotein subunit']</td>\n",
       "      <td>MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...</td>\n",
       "      <td>602</td>\n",
       "      <td>['PF00890', 'PF02910']</td>\n",
       "      <td>['ecj:JW4115', 'eco:b4154']</td>\n",
       "      <td>['NP_418578.1', 'NC_000913.3', 'WP_001192973.1...</td>\n",
       "      <td>['GO:0005829; C:cytosol; IDA:EcoCyc.', 'GO:001...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00370</th>\n",
       "      <td>P00370</td>\n",
       "      <td>gdhA</td>\n",
       "      <td>['1.4.1.4']</td>\n",
       "      <td>['NADP-GDH', 'NADP-specific glutamate dehydrog...</td>\n",
       "      <td>MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...</td>\n",
       "      <td>447</td>\n",
       "      <td>['PF00208', 'PF02812']</td>\n",
       "      <td>['ecj:JW1750', 'eco:b1761']</td>\n",
       "      <td>['NP_416275.1', 'NC_000913.3', 'WP_000373021.1...</td>\n",
       "      <td>['GO:0005737; C:cytoplasm; IDA:EcoliWiki.', 'G...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00393</th>\n",
       "      <td>P00393</td>\n",
       "      <td>ndh</td>\n",
       "      <td>['1.6.99.3']</td>\n",
       "      <td>['NADH dehydrogenase']</td>\n",
       "      <td>MTTPLKKIVIVGGGAGGLEMATQLGHKLGRKKKAKITLVDRNHSHL...</td>\n",
       "      <td>434</td>\n",
       "      <td>['PF07992']</td>\n",
       "      <td>['ecj:JW1095', 'eco:b1109']</td>\n",
       "      <td>['NP_415627.1', 'NC_000913.3', 'WP_000211045.1...</td>\n",
       "      <td>['GO:0030964; C:NADH dehydrogenase complex; ID...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       u_uniprot_acc u_gene_name   u_ec_number  \\\n",
       "O32583        O32583        thiS           NaN   \n",
       "P00350        P00350         gnd  ['1.1.1.44']   \n",
       "P00363        P00363        frdA   ['1.3.5.4']   \n",
       "P00370        P00370        gdhA   ['1.4.1.4']   \n",
       "P00393        P00393         ndh  ['1.6.99.3']   \n",
       "\n",
       "                                            u_description  \\\n",
       "O32583  ['Thiamine biosynthesis protein ThiS', 'Sulfur...   \n",
       "P00350  ['6-phosphogluconate dehydrogenase, decarboxyl...   \n",
       "P00363        ['Fumarate reductase flavoprotein subunit']   \n",
       "P00370  ['NADP-GDH', 'NADP-specific glutamate dehydrog...   \n",
       "P00393                             ['NADH dehydrogenase']   \n",
       "\n",
       "                                                    u_seq  u_seq_len  \\\n",
       "O32583  MQILFNDQAMQCAAGQTVHELLEQLDQRQAGAALAINQQIVPREQW...         66   \n",
       "P00350  MSKQQIGVVGMAVMGRNLALNIESRGYTVSIFNRSREKTEEVIAEN...        468   \n",
       "P00363  MQTFQADLAIVGAGGAGLRAAIAAAQANPNAKIALISKVYPMRSHT...        602   \n",
       "P00370  MDQTYSLESFLNHVQKRDPNQTEFAQAVREVMTTLWPFLEQNPKYR...        447   \n",
       "P00393  MTTPLKKIVIVGGGAGGLEMATQLGHKLGRKKKAKITLVDRNHSHL...        434   \n",
       "\n",
       "                        u_pfam                    u_kegg_id  \\\n",
       "O32583             ['PF02597']  ['ecj:JW3955', 'eco:b4407']   \n",
       "P00350  ['PF00393', 'PF03446']  ['ecj:JW2011', 'eco:b2029']   \n",
       "P00363  ['PF00890', 'PF02910']  ['ecj:JW4115', 'eco:b4154']   \n",
       "P00370  ['PF00208', 'PF02812']  ['ecj:JW1750', 'eco:b1761']   \n",
       "P00393             ['PF07992']  ['ecj:JW1095', 'eco:b1109']   \n",
       "\n",
       "                                                 u_refseq  \\\n",
       "O32583  ['WP_001166226.1', 'NZ_LN832404.1', 'YP_026279...   \n",
       "P00350  ['NP_416533.1', 'NC_000913.3', 'WP_000043484.1...   \n",
       "P00363  ['NP_418578.1', 'NC_000913.3', 'WP_001192973.1...   \n",
       "P00370  ['NP_416275.1', 'NC_000913.3', 'WP_000373021.1...   \n",
       "P00393  ['NP_415627.1', 'NC_000913.3', 'WP_000211045.1...   \n",
       "\n",
       "                                                     u_go u_reviewed  \n",
       "O32583  ['GO:0000166; F:nucleotide binding; IEA:UniPro...       True  \n",
       "P00350  ['GO:0005829; C:cytosol; IDA:EcoCyc.', 'GO:004...       True  \n",
       "P00363  ['GO:0005829; C:cytosol; IDA:EcoCyc.', 'GO:001...       True  \n",
       "P00370  ['GO:0005737; C:cytoplasm; IDA:EcoliWiki.', 'G...       True  \n",
       "P00393  ['GO:0030964; C:NADH dehydrogenase complex; ID...       True  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_METADATA_UNIPROT = pd.read_csv(os.path.join(DATA_FRAMES, 'DF_METADATA_UNIPROT.csv'), index_col=0)\n",
    "DF_METADATA_UNIPROT.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:RED\">**ABOVE DATAFRAME -- to be provided**</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtaining structure metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:orange\">**SKIP**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "os.chdir(STRUCT_EXP_FILES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pdb_ids = DF_MAP_GENE_UNIPROT_PDB[pd.notnull(DF_MAP_GENE_UNIPROT_PDB.u_pdb)].u_pdb.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***PROGRESS: PARSED & DOWNLOADED 3559/3559 PDB IDS***\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:.prody:Connecting wwPDB FTP server RCSB PDB (USA).\n",
      "DEBUG:.prody:1oro downloaded (1oro.pdb)\n",
      "DEBUG:.prody:PDB download via FTP completed (1 downloaded, 0 failed).\n",
      "DEBUG:.prody:PDB file is found in working directory (1oro.pdb).\n"
     ]
    }
   ],
   "source": [
    "pdb_metadata_all = pdb_metadata_and_download(pdb_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_METADATA_PDB = pd.DataFrame.from_dict(pdb_metadata_all, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_METADATA_PDB.to_csv(os.path.join(DATA_FRAMES,'DF_METADATA_PDB.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>p_experiment</th>\n",
       "      <th>p_doi</th>\n",
       "      <th>p_chemicals</th>\n",
       "      <th>p_chain_uniprot_map</th>\n",
       "      <th>p_chains</th>\n",
       "      <th>p_pmid</th>\n",
       "      <th>p_space_group</th>\n",
       "      <th>p_deposition_date</th>\n",
       "      <th>p_ec_numbers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11AS</th>\n",
       "      <td>2.5</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>10.1038/NSB0198-15</td>\n",
       "      <td>['ASN']</td>\n",
       "      <td>{'A': ['P00963'], 'B': ['P00963']}</td>\n",
       "      <td>['A', 'B']</td>\n",
       "      <td>9437423</td>\n",
       "      <td>P 1 21 1</td>\n",
       "      <td>02-DEC-97</td>\n",
       "      <td>[('A', ['6.3.1.1']), ('B', ['6.3.1.1'])]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12AS</th>\n",
       "      <td>2.2</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>10.1038/NSB0198-15</td>\n",
       "      <td>['AMP', 'ASN']</td>\n",
       "      <td>{'A': ['P00963'], 'B': ['P00963']}</td>\n",
       "      <td>['A', 'B']</td>\n",
       "      <td>9437423</td>\n",
       "      <td>P 1 21 1</td>\n",
       "      <td>02-DEC-97</td>\n",
       "      <td>[('A', ['6.3.1.1']), ('B', ['6.3.1.1'])]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>SOLUTION NMR</td>\n",
       "      <td>10.1021/BI980136Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'A': ['P0AEG4']}</td>\n",
       "      <td>['A']</td>\n",
       "      <td>9572841</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-98</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>SOLUTION NMR</td>\n",
       "      <td>10.1021/BI980136Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'A': ['P0AEG4']}</td>\n",
       "      <td>['A']</td>\n",
       "      <td>9572841</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-98</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1A2J</th>\n",
       "      <td>2.0</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>10.1016/S0969-2126(98)00077-X</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'A': ['P0AEG4']}</td>\n",
       "      <td>['A']</td>\n",
       "      <td>9655827</td>\n",
       "      <td>P 1 21 1</td>\n",
       "      <td>06-JAN-98</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      p_resolution       p_experiment                          p_doi  \\\n",
       "11AS           2.5  X-RAY DIFFRACTION             10.1038/NSB0198-15   \n",
       "12AS           2.2  X-RAY DIFFRACTION             10.1038/NSB0198-15   \n",
       "1A23           NaN       SOLUTION NMR              10.1021/BI980136Y   \n",
       "1A24           NaN       SOLUTION NMR              10.1021/BI980136Y   \n",
       "1A2J           2.0  X-RAY DIFFRACTION  10.1016/S0969-2126(98)00077-X   \n",
       "\n",
       "         p_chemicals                 p_chain_uniprot_map    p_chains   p_pmid  \\\n",
       "11AS         ['ASN']  {'A': ['P00963'], 'B': ['P00963']}  ['A', 'B']  9437423   \n",
       "12AS  ['AMP', 'ASN']  {'A': ['P00963'], 'B': ['P00963']}  ['A', 'B']  9437423   \n",
       "1A23             NaN                   {'A': ['P0AEG4']}       ['A']  9572841   \n",
       "1A24             NaN                   {'A': ['P0AEG4']}       ['A']  9572841   \n",
       "1A2J             NaN                   {'A': ['P0AEG4']}       ['A']  9655827   \n",
       "\n",
       "     p_space_group p_deposition_date                              p_ec_numbers  \n",
       "11AS      P 1 21 1         02-DEC-97  [('A', ['6.3.1.1']), ('B', ['6.3.1.1'])]  \n",
       "12AS      P 1 21 1         02-DEC-97  [('A', ['6.3.1.1']), ('B', ['6.3.1.1'])]  \n",
       "1A23           NaN         15-JAN-98                                       NaN  \n",
       "1A24           NaN         15-JAN-98                                       NaN  \n",
       "1A2J      P 1 21 1         06-JAN-98                                       NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_METADATA_PDB = pd.read_csv(os.path.join(DATA_FRAMES,'DF_METADATA_PDB.csv'), index_col=0)\n",
    "DF_METADATA_PDB.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:RED\">**ABOVE DATAFRAME -- to be provided**</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final integrated DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:#66FF99\">**The last step of the ID mapping process is to organize the information by reaction, according to the model.**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "# creating the final data frame organized according to reaction\n",
    "DF_01_RXN_GENE_UNIPROT_PDB = pd.DataFrame()\n",
    "counter = 1\n",
    "\n",
    "for rxn_object in tqdm(model.reactions):\n",
    "    \n",
    "    m_reaction = rxn_object.id\n",
    "    m_gene_reaction_rule = rxn_object.gene_reaction_rule\n",
    "    m_genes = rxn_object.genes\n",
    "    metabolites = rxn_object.metabolites\n",
    "    m_metabolites = [x.id for x in metabolites]\n",
    "    m_subsystem = rxn_object.subsystem\n",
    "    m_formula = rxn_object.reaction\n",
    "\n",
    "    if len(m_genes) == 0:\n",
    "        continue\n",
    "\n",
    "    for m_gene in m_genes:        \n",
    "        extract = DF_MAP_GENE_UNIPROT_PDB[DF_MAP_GENE_UNIPROT_PDB.m_gene == m_gene.id]\n",
    "        u_uniprot_accs = list(extract[extract.u_uniprot_acc.notnull()].u_uniprot_acc.unique())\n",
    "        \n",
    "        if m_gene.id in final_missing:\n",
    "            append_df = {}\n",
    "            append_df['m_reaction'] = m_reaction\n",
    "            append_df['m_formula'] = m_formula\n",
    "            append_df['m_gene_reaction_rule'] = m_gene_reaction_rule\n",
    "            append_df['m_metabolites'] = m_metabolites\n",
    "            append_df['m_subsystem'] = m_subsystem\n",
    "            append_df['m_gene'] = m_gene.id\n",
    "            DF_01_RXN_GENE_UNIPROT_PDB = DF_01_RXN_GENE_UNIPROT_PDB.append(append_df, ignore_index = True)\n",
    "            \n",
    "        else:\n",
    "            for u_uniprot_acc in u_uniprot_accs:\n",
    "                u_pdb_list = DF_MAP_GENE_UNIPROT_PDB[DF_MAP_GENE_UNIPROT_PDB['u_uniprot_acc'] == u_uniprot_acc]['u_pdb'].values\n",
    "                if pd.isnull(u_pdb_list.all()):\n",
    "                    u_pdb_count = 0\n",
    "                else:\n",
    "                    u_pdb_count = len(set(u_pdb_list))\n",
    "\n",
    "\n",
    "                uniprot_metadata = DF_METADATA_UNIPROT.loc[u_uniprot_acc]\n",
    "\n",
    "                for u_pdb in u_pdb_list:\n",
    "                    append_df = {}\n",
    "                    append_df['m_reaction'] = m_reaction\n",
    "                    append_df['m_formula'] = m_formula\n",
    "                    append_df['m_gene_reaction_rule'] = m_gene_reaction_rule\n",
    "                    append_df['m_metabolites'] = m_metabolites\n",
    "                    append_df['m_subsystem'] = m_subsystem\n",
    "                    append_df['m_gene'] = m_gene.id\n",
    "\n",
    "                    append_df.update(uniprot_metadata)\n",
    "                    append_df['u_pdb_count'] = u_pdb_count\n",
    "\n",
    "                    # Todo: what ? is this a copy paste error?\n",
    "                    if u_pdb_count != u_pdb in DF_METADATA_PDB.index:\n",
    "                        append_df['u_pdb'] = u_pdb\n",
    "                        pdb_metadata = DF_METADATA_PDB.loc[u_pdb]\n",
    "                        append_df.update(pdb_metadata)\n",
    "\n",
    "                    DF_01_RXN_GENE_UNIPROT_PDB = DF_01_RXN_GENE_UNIPROT_PDB.append(append_df, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ordered_props = ['m_reaction','m_subsystem','m_formula','m_metabolites','m_gene_reaction_rule','m_gene',\n",
    "'u_uniprot_acc','u_reviewed','u_gene_name','u_ec_number','u_description','u_seq','u_seq_len',\n",
    "'u_pfam','u_kegg_id','u_refseq','u_go','u_pdb_count','u_pdb',\n",
    "'p_experiment','p_resolution','p_chemicals','p_chains','p_chain_uniprot_map',\n",
    "'p_ec_numbers','p_deposition_date','p_doi','p_pmid','p_space_group']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_reaction</th>\n",
       "      <th>m_subsystem</th>\n",
       "      <th>m_formula</th>\n",
       "      <th>m_metabolites</th>\n",
       "      <th>m_gene_reaction_rule</th>\n",
       "      <th>m_gene</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>u_gene_name</th>\n",
       "      <th>u_ec_number</th>\n",
       "      <th>u_description</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_pfam</th>\n",
       "      <th>u_kegg_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_go</th>\n",
       "      <th>u_pdb_count</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th>p_experiment</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>p_chemicals</th>\n",
       "      <th>p_chains</th>\n",
       "      <th>p_chain_uniprot_map</th>\n",
       "      <th>p_ec_numbers</th>\n",
       "      <th>p_deposition_date</th>\n",
       "      <th>p_doi</th>\n",
       "      <th>p_pmid</th>\n",
       "      <th>p_space_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>1BT9</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02931']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01-SEP-98</td>\n",
       "      <td>10.1021/BI981215C</td>\n",
       "      <td>9843370</td>\n",
       "      <td>P 3 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>4JFB</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A', 'C', 'B', 'F', 'D', 'E']</td>\n",
       "      <td>{'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28-FEB-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C 1 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>3FYX</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.40</td>\n",
       "      <td>['EPE', '451']</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02931']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23-JAN-09</td>\n",
       "      <td>10.1002/ANIE.200900457</td>\n",
       "      <td>19322865</td>\n",
       "      <td>P 3 2 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>3K1B</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>4.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A', 'C', 'B', 'D']</td>\n",
       "      <td>{'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26-SEP-09</td>\n",
       "      <td>10.1002/PRO.369</td>\n",
       "      <td>20196071</td>\n",
       "      <td>P 3 2 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   m_reaction                      m_subsystem  \\\n",
       "37  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "5   12PPDRtex  Transport, Outer Membrane Porin   \n",
       "6   12PPDRtex  Transport, Outer Membrane Porin   \n",
       "7   12PPDRtex  Transport, Outer Membrane Porin   \n",
       "8   12PPDRtex  Transport, Outer Membrane Porin   \n",
       "\n",
       "                            m_formula                     m_metabolites  \\\n",
       "37  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "5   12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "6   12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "7   12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "8   12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "\n",
       "                  m_gene_reaction_rule m_gene u_uniprot_acc  u_reviewed  \\\n",
       "37  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "5   (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "6   (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "7   (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "8   (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "\n",
       "   u_gene_name u_ec_number                                      u_description  \\\n",
       "37        phoE         NaN                  ['Outer membrane pore protein E']   \n",
       "5         ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "6         ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "7         ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "8         ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "\n",
       "                                                u_seq  u_seq_len       u_pfam  \\\n",
       "37  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "5   MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "6   MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "7   MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "8   MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "\n",
       "                      u_kegg_id  \\\n",
       "37  ['ecj:JW0231', 'eco:b0241']   \n",
       "5   ['ecj:JW0912', 'eco:b0929']   \n",
       "6   ['ecj:JW0912', 'eco:b0929']   \n",
       "7   ['ecj:JW0912', 'eco:b0929']   \n",
       "8   ['ecj:JW0912', 'eco:b0929']   \n",
       "\n",
       "                                             u_refseq  \\\n",
       "37  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "5   ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "6   ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "7   ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "8   ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "\n",
       "                                                 u_go  u_pdb_count u_pdb  \\\n",
       "37  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "5   ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  1BT9   \n",
       "6   ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  4JFB   \n",
       "7   ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  3FYX   \n",
       "8   ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  3K1B   \n",
       "\n",
       "         p_experiment  p_resolution     p_chemicals  \\\n",
       "37  X-RAY DIFFRACTION          3.00             NaN   \n",
       "5   X-RAY DIFFRACTION          3.00             NaN   \n",
       "6   X-RAY DIFFRACTION          3.80             NaN   \n",
       "7   X-RAY DIFFRACTION          3.40  ['EPE', '451']   \n",
       "8   X-RAY DIFFRACTION          4.39             NaN   \n",
       "\n",
       "                          p_chains  \\\n",
       "37                           ['A']   \n",
       "5                            ['A']   \n",
       "6   ['A', 'C', 'B', 'F', 'D', 'E']   \n",
       "7                            ['A']   \n",
       "8             ['A', 'C', 'B', 'D']   \n",
       "\n",
       "                                  p_chain_uniprot_map p_ec_numbers  \\\n",
       "37                                  {'A': ['P02932']}          NaN   \n",
       "5                                   {'A': ['P02931']}          NaN   \n",
       "6   {'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...          NaN   \n",
       "7                                   {'A': ['P02931']}          NaN   \n",
       "8   {'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...          NaN   \n",
       "\n",
       "   p_deposition_date                   p_doi    p_pmid p_space_group  \n",
       "37         15-JAN-93        10.1038/358727A0   1380671       P 3 2 1  \n",
       "5          01-SEP-98       10.1021/BI981215C   9843370       P 3 2 1  \n",
       "6          28-FEB-13                     NaN       NaN       C 1 2 1  \n",
       "7          23-JAN-09  10.1002/ANIE.200900457  19322865       P 3 2 1  \n",
       "8          26-SEP-09         10.1002/PRO.369  20196071       P 3 2 1  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_01_RXN_GENE_UNIPROT_PDB = DF_01_RXN_GENE_UNIPROT_PDB[ordered_props].reset_index(drop=True).sort_values(by=['m_reaction','m_gene'])\n",
    "DF_01_RXN_GENE_UNIPROT_PDB.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_01_RXN_GENE_UNIPROT_PDB.to_pickle(os.path.join(DATA_FRAMES, 'DF_01_RXN_GENE_UNIPROT_PDB.pckl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mapping to homology models\n",
    "\n",
    "<span style=\"background-color:#66FF99\">**Next, we use homology models to represent structures that do not have an experimental structure available.**</span>\n",
    "\n",
    "We have already generated the homology models for this model, thus, we just need to indicate their existence in the GEM-PRO dataframe. All homology models will be stored in the folder:\n",
    "\n",
    "    structure_files/homology_models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading existing gene -> homology model dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_MAP_HOMOLOGY = pd.read_csv(os.path.join(DATA_FRAMES, 'DF_MAP_HOMOLOGY.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>i_entry_name</th>\n",
       "      <th>i_length</th>\n",
       "      <th>i_native</th>\n",
       "      <th>i_tm_helix</th>\n",
       "      <th>i_tm_score</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAEX_ECOLI</td>\n",
       "      <td>67</td>\n",
       "      <td></td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P46478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALPA_ECOLI</td>\n",
       "      <td>70</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P33997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARNE_ECOLI</td>\n",
       "      <td>111</td>\n",
       "      <td></td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q47377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARSR_ECOLI</td>\n",
       "      <td>117</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P37309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BFD_ECOLI</td>\n",
       "      <td>64</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P0AE56</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  i_entry_name  i_length i_native  i_tm_helix  i_tm_score u_uniprot_acc\n",
       "0   AAEX_ECOLI        67                   36         NaN        P46478\n",
       "1   ALPA_ECOLI        70                    0         NaN        P33997\n",
       "2   ARNE_ECOLI       111                   50         NaN        Q47377\n",
       "3   ARSR_ECOLI       117                    0         NaN        P37309\n",
       "4    BFD_ECOLI        64                    0         NaN        P0AE56"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_MAP_HOMOLOGY.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging with the created GEM-PRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DF_02_RXN_GENE_UNIPROT_PDB_HOMOL = DF_01_RXN_GENE_UNIPROT_PDB.merge(DF_MAP_HOMOLOGY, how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"background-color:#66FF99\">**Note the addition of homology model info to the right.**</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_reaction</th>\n",
       "      <th>m_subsystem</th>\n",
       "      <th>m_formula</th>\n",
       "      <th>m_metabolites</th>\n",
       "      <th>m_gene_reaction_rule</th>\n",
       "      <th>m_gene</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>u_gene_name</th>\n",
       "      <th>u_ec_number</th>\n",
       "      <th>u_description</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_pfam</th>\n",
       "      <th>u_kegg_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_go</th>\n",
       "      <th>u_pdb_count</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th>p_experiment</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>p_chemicals</th>\n",
       "      <th>p_chains</th>\n",
       "      <th>p_chain_uniprot_map</th>\n",
       "      <th>p_ec_numbers</th>\n",
       "      <th>p_deposition_date</th>\n",
       "      <th>p_doi</th>\n",
       "      <th>p_pmid</th>\n",
       "      <th>p_space_group</th>\n",
       "      <th>i_entry_name</th>\n",
       "      <th>i_length</th>\n",
       "      <th>i_native</th>\n",
       "      <th>i_tm_helix</th>\n",
       "      <th>i_tm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>1BT9</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02931']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>01-SEP-98</td>\n",
       "      <td>10.1021/BI981215C</td>\n",
       "      <td>9843370</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>OMPF_ECOLI</td>\n",
       "      <td>362</td>\n",
       "      <td>1opfA</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>4JFB</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A', 'C', 'B', 'F', 'D', 'E']</td>\n",
       "      <td>{'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28-FEB-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C 1 2 1</td>\n",
       "      <td>OMPF_ECOLI</td>\n",
       "      <td>362</td>\n",
       "      <td>1opfA</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>3FYX</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3.40</td>\n",
       "      <td>['EPE', '451']</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02931']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23-JAN-09</td>\n",
       "      <td>10.1002/ANIE.200900457</td>\n",
       "      <td>19322865</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>OMPF_ECOLI</td>\n",
       "      <td>362</td>\n",
       "      <td>1opfA</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0929</td>\n",
       "      <td>P02931</td>\n",
       "      <td>1</td>\n",
       "      <td>ompF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Porin OmpF', 'Outer membrane protein IA', 'O...</td>\n",
       "      <td>MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...</td>\n",
       "      <td>362</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0912', 'eco:b0929']</td>\n",
       "      <td>['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IDA:EcoCy...</td>\n",
       "      <td>32</td>\n",
       "      <td>3K1B</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>4.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A', 'C', 'B', 'D']</td>\n",
       "      <td>{'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26-SEP-09</td>\n",
       "      <td>10.1002/PRO.369</td>\n",
       "      <td>20196071</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>OMPF_ECOLI</td>\n",
       "      <td>362</td>\n",
       "      <td>1opfA</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  m_reaction                      m_subsystem  \\\n",
       "0  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "1  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "2  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "3  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "4  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "\n",
       "                           m_formula                     m_metabolites  \\\n",
       "0  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "1  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "2  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "3  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "4  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "\n",
       "                 m_gene_reaction_rule m_gene u_uniprot_acc  u_reviewed  \\\n",
       "0  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "1  (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "2  (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "3  (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "4  (b0241 or b0929 or b1377 or b2215)  b0929        P02931           1   \n",
       "\n",
       "  u_gene_name u_ec_number                                      u_description  \\\n",
       "0        phoE         NaN                  ['Outer membrane pore protein E']   \n",
       "1        ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "2        ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "3        ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "4        ompF         NaN  ['Porin OmpF', 'Outer membrane protein IA', 'O...   \n",
       "\n",
       "                                               u_seq  u_seq_len       u_pfam  \\\n",
       "0  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "1  MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "2  MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "3  MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "4  MMKRNILAVIVPALLVAGTANAAEIYNKDGNKVDLYGKAVGLHYFS...        362  ['PF00267']   \n",
       "\n",
       "                     u_kegg_id  \\\n",
       "0  ['ecj:JW0231', 'eco:b0241']   \n",
       "1  ['ecj:JW0912', 'eco:b0929']   \n",
       "2  ['ecj:JW0912', 'eco:b0929']   \n",
       "3  ['ecj:JW0912', 'eco:b0929']   \n",
       "4  ['ecj:JW0912', 'eco:b0929']   \n",
       "\n",
       "                                            u_refseq  \\\n",
       "0  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "1  ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "2  ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "3  ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "4  ['NP_415449.1', 'NC_000913.3', 'WP_000977920.1...   \n",
       "\n",
       "                                                u_go  u_pdb_count u_pdb  \\\n",
       "0  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "1  ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  1BT9   \n",
       "2  ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  4JFB   \n",
       "3  ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  3FYX   \n",
       "4  ['GO:0009279; C:cell outer membrane; IDA:EcoCy...           32  3K1B   \n",
       "\n",
       "        p_experiment  p_resolution     p_chemicals  \\\n",
       "0  X-RAY DIFFRACTION          3.00             NaN   \n",
       "1  X-RAY DIFFRACTION          3.00             NaN   \n",
       "2  X-RAY DIFFRACTION          3.80             NaN   \n",
       "3  X-RAY DIFFRACTION          3.40  ['EPE', '451']   \n",
       "4  X-RAY DIFFRACTION          4.39             NaN   \n",
       "\n",
       "                         p_chains  \\\n",
       "0                           ['A']   \n",
       "1                           ['A']   \n",
       "2  ['A', 'C', 'B', 'F', 'D', 'E']   \n",
       "3                           ['A']   \n",
       "4            ['A', 'C', 'B', 'D']   \n",
       "\n",
       "                                 p_chain_uniprot_map p_ec_numbers  \\\n",
       "0                                  {'A': ['P02932']}          NaN   \n",
       "1                                  {'A': ['P02931']}          NaN   \n",
       "2  {'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...          NaN   \n",
       "3                                  {'A': ['P02931']}          NaN   \n",
       "4  {'A': ['P02931'], 'C': ['P02931'], 'B': ['P029...          NaN   \n",
       "\n",
       "  p_deposition_date                   p_doi    p_pmid p_space_group  \\\n",
       "0         15-JAN-93        10.1038/358727A0   1380671       P 3 2 1   \n",
       "1         01-SEP-98       10.1021/BI981215C   9843370       P 3 2 1   \n",
       "2         28-FEB-13                     NaN       NaN       C 1 2 1   \n",
       "3         23-JAN-09  10.1002/ANIE.200900457  19322865       P 3 2 1   \n",
       "4         26-SEP-09         10.1002/PRO.369  20196071       P 3 2 1   \n",
       "\n",
       "  i_entry_name  i_length i_native  i_tm_helix  i_tm_score  \n",
       "0   PHOE_ECOLI       351    1phoA           7         NaN  \n",
       "1   OMPF_ECOLI       362    1opfA          13         NaN  \n",
       "2   OMPF_ECOLI       362    1opfA          13         NaN  \n",
       "3   OMPF_ECOLI       362    1opfA          13         NaN  \n",
       "4   OMPF_ECOLI       362    1opfA          13         NaN  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.to_pickle(os.path.join(DATA_FRAMES, 'DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.pckl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ranking experimental structures\n",
    "\n",
    "<span style=\"background-color:#66FF99\">**Finally, we'd like to understand how well our experimental structures represent the gene within the model.**</span>\n",
    "\n",
    "We will demonstrate how we rank models based on two simple criteria:\n",
    "\n",
    "1. Sequence identity to the \"canonical\" amino acid sequence\n",
    "2. Structure resolution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Scoring sequence identity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To score sequence identity, we simply align the \"canonical\" amino acid sequence, which is the sequence obtained by the UniProt entry, to the available experimental structures. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing FASTA files for the original sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "uniprot_written_fasta_files = [os.path.basename(path) for path in glob.glob(SEQ_UNIPROT_FILES + '*.faa')]\n",
    "all_uniprot_ids = DF_02_RXN_GENE_UNIPROT_PDB_HOMOL[pd.notnull(DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.u_uniprot_acc)].u_uniprot_acc.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "os.chdir(SEQ_UNIPROT_FILES)\n",
    "\n",
    "counter = 1\n",
    "for uniprot_id in tqdm(all_uniprot_ids):\n",
    "    u_seq = DF_02_RXN_GENE_UNIPROT_PDB_HOMOL[DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.u_uniprot_acc == uniprot_id]['u_seq'].unique()[0]\n",
    "    u_seq_len = len(u_seq)\n",
    "    \n",
    "    if uniprot_id + '.faa' not in uniprot_written_fasta_files:\n",
    "        uniprot_seq = SeqRecord(Seq(u_seq, IUPAC.protein),id=uniprot_id,description='uniprot sequence')\n",
    "        write_fasta_file(uniprot_seq, uniprot_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing FASTA files for the structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pdb_written_fasta_files = [os.path.basename(path) for path in glob.glob(SEQ_PDB_FILES + '*.faa')]\n",
    "all_pdb_ids = DF_02_RXN_GENE_UNIPROT_PDB_HOMOL[pd.notnull(DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.u_pdb)].u_pdb.unique().tolist()\n",
    "all_pdb_ids = [pdb.lower() for pdb in all_pdb_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_pdb_seq(pdb_id, pdb_file):\n",
    "    '''\n",
    "    this uses biopython to get the sequence of the residues that are resolved in the structure. \n",
    "    IT ADDS AN X IN POSITIONS THAT HAVE NO RESOLVED RESIDUE!\n",
    "    IT EVEN WORKS WITH CIF FILES!\n",
    "    '''\n",
    "    \n",
    "    if pdb_file.endswith('.pdb'):\n",
    "        parser = PDBParser()\n",
    "        structure = parser.get_structure(pdb_id, pdb_file)\n",
    "    elif pdb_file.endswith('.cif'):\n",
    "        parser = MMCIFParser()\n",
    "        try:\n",
    "            structure = parser.get_structure(pdb_id, pdb_file)\n",
    "        except:\n",
    "            return None\n",
    "        \n",
    "    structure_seqs = []\n",
    "    for chain in structure[0]:\n",
    "        chain_seq = SeqRecord(Seq('', IUPAC.protein), name=pdb_id, id=pdb_id+'.'+chain.get_id(), description='PDB structure residues')\n",
    "        first = True\n",
    "        tracker = 0\n",
    "        for res in chain.get_residues():\n",
    "            if is_aa(res, standard=True):\n",
    "                full_id = res.get_full_id()\n",
    "                end_tracker = full_id[3][1]\n",
    "                i_code = full_id[3][2]\n",
    "                aa = three_to_one(res.get_resname())\n",
    "                \n",
    "                if end_tracker != (tracker + 1) and first == False:\n",
    "                    if i_code != ' ':\n",
    "                        chain_seq += aa\n",
    "                        tracker = end_tracker + 1\n",
    "                        continue\n",
    "\n",
    "                    else:\n",
    "                        chain_seq += 'X'*(end_tracker - tracker - 1)\n",
    "                        \n",
    "                chain_seq += aa\n",
    "                first = False\n",
    "                tracker = end_tracker\n",
    "\n",
    "        structure_seqs.append(chain_seq)\n",
    "\n",
    "    return structure_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pdb_seq_errors = []\n",
    "\n",
    "counter = 1\n",
    "for pdb_id in tqdm(all_pdb_ids):\n",
    "    \n",
    "    if pdb_id + '.faa' not in pdb_written_fasta_files:\n",
    "        os.chdir(STRUCT_EXP_FILES)\n",
    "        \n",
    "        try:\n",
    "            pdb_file = pdb_id + '.pdb'\n",
    "            pdb_sequences = get_pdb_seq(pdb_id, pdb_file)\n",
    "        except IOError:\n",
    "            pdb_file = pdb_id + '.cif'\n",
    "            print pdb_file\n",
    "            pdb_sequences = get_pdb_seq(pdb_id, pdb_file)\n",
    "            \n",
    "        if pdb_sequences == None:\n",
    "            warnings.warn('***ERROR: could not extract sequence for %s***' % pdb_id)\n",
    "            pdb_seq_errors.append(pdb_id)\n",
    "            continue\n",
    "        else:\n",
    "            os.chdir(SEQ_PDB_FILES)\n",
    "            write_fasta_file(pdb_sequences, pdb_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aligning the structures to the sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When aligning the structure to the sequence, there is one main consideration that must be handled. A structure can consist of multiple chains, which may be repeats of the same sequence (homo-mers) or may be transcripts of other genes (hetero-mers). We just want to score based on the chain which corresponds to the gene of interest. Thus, we take the chain that aligns with the highest sequence identity as our measure of % identity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tmp = DF_02_RXN_GENE_UNIPROT_PDB_HOMOL[pd.notnull(DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.u_pdb)]\n",
    "for_scoring = tmp[['u_uniprot_acc', 'u_pdb', 'u_seq_len', 'p_resolution']].drop_duplicates().set_index(['u_uniprot_acc', 'u_pdb']).sortlevel(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  1%|▏         | 8/612 [00:00<00:10, 56.00it/s]/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:32 UserWarning: ***WARNING: Previous error for 4v40. Was the PDB file able to be parsed?***\n",
      " 35%|███▍      | 213/612 [00:01<00:02, 141.10it/s]/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:50 UserWarning: ***ERROR: with alignment of P0ABA6 and 1d8s***\n",
      "/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:50 UserWarning: ***ERROR: with alignment of P0ABB0 and 1d8s***\n",
      "/usr/local/lib/python2.7/dist-packages/ipykernel/__main__.py:50 UserWarning: ***ERROR: with alignment of P0ABB4 and 1d8s***\n"
     ]
    }
   ],
   "source": [
    "os.chdir(SEQ_ALIGN_FILES)\n",
    "\n",
    "for_scoring['ssb_p_percent_seq_ident'] = np.nan\n",
    "\n",
    "uniprot_pdb_align_errors = []\n",
    "\n",
    "# looping through the dataframe using a multi index\n",
    "for uniprot_id in tqdm(for_scoring.index.get_level_values(0).unique()):\n",
    "    \n",
    "    alignment_scores = {}\n",
    "    alignment_coverage = {}\n",
    "    alignment_coverage_sim = {}\n",
    "\n",
    "    pdb_ids = for_scoring.loc[uniprot_id].index.unique().tolist()\n",
    "    \n",
    "    # if there are no PDBs associated with a UniProt ID, skip this UniProt ID\n",
    "    if np.nan in pdb_ids:\n",
    "        continue\n",
    "        \n",
    "    pdb_ids = [pdb_id.lower() for pdb_id in pdb_ids]\n",
    "    \n",
    "    uniprot_len = for_scoring.loc[uniprot_id].u_seq_len.unique()[0]\n",
    "    \n",
    "    uniprot_fasta = os.path.join(SEQ_UNIPROT_FILES, uniprot_id + '.faa')\n",
    "    if not os.path.exists(uniprot_fasta):\n",
    "        warnings.warn('***WARNING: No UniProt FASTA file found for %s.***' % uniprot_id)\n",
    "        uniprot_pdb_align_errors.append(uniprot_id)\n",
    "        continue\n",
    "    \n",
    "    for pdb_id in pdb_ids:\n",
    "        if pdb_id in pdb_seq_errors:\n",
    "            warnings.warn('***WARNING: Previous error for %s. Was the PDB file able to be parsed?***' % pdb_id)\n",
    "            uniprot_pdb_align_errors.append((uniprot_id, pdb_id))\n",
    "            continue\n",
    "            \n",
    "        pdb_fasta = os.path.join(SEQ_PDB_FILES, pdb_id + '.faa')\n",
    "        if not os.path.exists(pdb_fasta):\n",
    "            warnings.warn('***WARNING: No PDB FASTA file found for %s.***' % pdb_id)\n",
    "            uniprot_pdb_align_errors.append((uniprot_id, pdb_id))\n",
    "            continue\n",
    "            \n",
    "        # Run the alignment\n",
    "        alignment_filename = \"%s_%s_align.txt\" % (uniprot_id, pdb_id)\n",
    "        if not os.path.exists(os.path.join(SEQ_ALIGN_FILES, alignment_filename)):\n",
    "            try:\n",
    "                # print 'Running alignment'\n",
    "                alignment_filename = run_alignment(uniprot_id, uniprot_fasta, pdb_id, pdb_fasta)\n",
    "            except:\n",
    "                uniprot_pdb_align_errors.append((uniprot_id,pdb_id))\n",
    "                warnings.warn('***ERROR: with alignment of %s and %s***' % (uniprot_id, pdb_id))\n",
    "                alignment_scores[pdb_id] = 0\n",
    "                alignment_coverage[pdb_id] = 0\n",
    "                alignment_coverage_sim[pdb_id] = 0\n",
    "                continue\n",
    "\n",
    "        # Parse the alignment\n",
    "        os.chdir(SEQ_ALIGN_FILES)\n",
    "        alignment_properties = needle_reader(alignment_filename)\n",
    "\n",
    "        # Choose the best aligning chain\n",
    "        best_score = 0\n",
    "        for pdb_alignment_props in alignment_properties.values():\n",
    "            if pdb_alignment_props['score'] > best_score:\n",
    "                best_score = pdb_alignment_props['score']\n",
    "\n",
    "        best_identity = 0\n",
    "        for pdb_alignment_props in alignment_properties.values():\n",
    "            if pdb_alignment_props['identity'] > best_identity:\n",
    "                best_identity = pdb_alignment_props['identity']\n",
    "        \n",
    "        best_similarity = 0\n",
    "        for pdb_alignment_props in alignment_properties.values():\n",
    "            if pdb_alignment_props['similarity'] > best_similarity:\n",
    "                best_similarity = pdb_alignment_props['similarity']\n",
    "\n",
    "        alignment_scores[pdb_id] = best_score\n",
    "        alignment_coverage[pdb_id] = best_identity\n",
    "        alignment_coverage_sim[pdb_id] = best_similarity\n",
    "\n",
    "    for pdb, cov in alignment_coverage.iteritems():\n",
    "        # NOTE: setting the value by .loc in a multi indexed dataframe REQUIRES IT TO BE SORTED\n",
    "        # http://stackoverflow.com/questions/16833842/assign-new-values-to-slice-from-multiindex-dataframe\n",
    "        for_scoring.loc[(uniprot_id, pdb.upper()), 'ssb_p_aln_coverage'] = cov\n",
    "        for_scoring.loc[(uniprot_id, pdb.upper()), 'ssb_p_aln_coverage_sim'] = alignment_coverage_sim[pdb]\n",
    "        for_scoring.loc[(uniprot_id, pdb.upper()), 'ssb_p_percent_seq_ident'] = float(cov) / float(uniprot_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">O32583</th>\n",
       "      <th>1F0Z</th>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1ZUD</th>\n",
       "      <td>66</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">P00350</th>\n",
       "      <th>2ZYA</th>\n",
       "      <td>468</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2ZYD</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3FWN</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.995726</td>\n",
       "      <td>466</td>\n",
       "      <td>466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     u_seq_len  p_resolution  ssb_p_percent_seq_ident  \\\n",
       "u_uniprot_acc u_pdb                                                     \n",
       "O32583        1F0Z          66           NaN                 1.000000   \n",
       "              1ZUD          66          1.98                 1.000000   \n",
       "P00350        2ZYA         468          1.60                 0.993590   \n",
       "              2ZYD         468          1.50                 0.993590   \n",
       "              3FWN         468          1.50                 0.995726   \n",
       "\n",
       "                     ssb_p_aln_coverage  ssb_p_aln_coverage_sim  \n",
       "u_uniprot_acc u_pdb                                              \n",
       "O32583        1F0Z                   66                      66  \n",
       "              1ZUD                   66                      66  \n",
       "P00350        2ZYA                  465                     465  \n",
       "              2ZYD                  465                     465  \n",
       "              3FWN                  466                     466  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_scoring.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for_scoring = for_scoring[pd.notnull(for_scoring.ssb_p_percent_seq_ident)]\n",
    "seq_ident = for_scoring.ssb_p_percent_seq_ident"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    3875.000000\n",
      "mean        0.916043\n",
      "std         0.165510\n",
      "min         0.000000\n",
      "10%         0.827133\n",
      "25%         0.924012\n",
      "50%         0.973199\n",
      "75%         0.993464\n",
      "90%         0.997658\n",
      "max         1.000000\n",
      "Name: ssb_p_percent_seq_ident, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "seq_ident_stats = seq_ident.describe(percentiles = [.1,.25,.5,.75,.9])\n",
    "seq_ident_stats_simple_cutoff_10 = seq_ident_stats['10%']\n",
    "seq_ident_stats_simple_cutoff_25 = seq_ident_stats['25%']\n",
    "print seq_ident_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.4, 1.1)"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeEAAAEjCAYAAAAFarmxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmYVPWd7/H3qa2r925EWhBQtqMgm4AICooSI7nXm3HM\n+JhMYhxjktEkRn3MdebeTCaTmbl6k5h7zehk8Y4ZdcZRx8TEJW4oigIisskqB1AQoelm6X2r7dw/\nTlVTtI10N1X1q+r+vJ6nnlq66pzvr6q6Pmf5/c4BEREREREREREREREREREREREREREREREREZEB\nszI1Idd13UxNS0REJN9ZlnXKGerLRCEiIiLSfwphERERQxTCIiIihiiERUREDFEIi4iIGKIQFhER\nMUQhLCIiYohCWERExBCFsIiIiCEKYREREUMUwiIiIoYohEVERAxRCIuIiBiiEBYRETFEISwiImKI\nQlhERMQQhbCIiIghCmERERFDFMIiIiKGKIRFREQMUQiLiIgYohAWERExJGC6ABERyQ+dnZ3s3LmT\nUCjEOeecY7qcIUEhLCIyREUiEXbu3Mn27dvZvn07u3fvJh6PA3DNNddw9dVXY1mW4SoHt4y9u67r\nupmaloiIZEcikWDp0qWsW7eOnbt2EYtG0/5qEagcRqz5KLguCxYs4KabbiIQ0Ppab6wMLKEohEVE\nhpDHHnuMl156qft+oPI0gqeP9C7DR+ILhuiq3UvLmtdx4zEmT57MbbfdRmlpqcGq85NCWERE+uyF\nF17g8ccfB5+P8pkLCI0ci68o3Otzow2HaV71MomuDkaOHMn3vvc9RowYkeOK85tCWERE+mTVqlX8\n8pe/BKD8gssIj5lw0tfE21tpWvUy8eYGKioquOOOO5g4cWK2Sy0YCmERETmprVu38tOf/pR4PE7p\ntLmUTJre59cmohGa33mNaP1+gsEgt9xyCxdccEEWqy0cmQhhjRMWERnE9uzZw3333Uc8Hqd44tR+\nBTCALxii8qIrCZ99DtFolPvvv5+tW7dmqdqhRyEsIjJI1dfXc++999LZ2UnR6AmUTrtwQNOxfD7K\nzl9A8aRpuK7Lk08+iTZ+ZoZCWERkEGppaeEnP/kJTU1NBE8fRfnsS05pzK9lWZROno2vqJgPP/yQ\ntWvXZrDaoUshLCIyyHR1dfGzn/2Muro6/JXDqJj3GSy//5SnawUClJx7PgC/+93vSCQSpzzNoU4h\nLCIyyDzxxBPs3r0bX0kZlRctwRcMZWza4XHn4CspY//+/axatSpj0x2qFMIiIoPIzp07ee2118Dy\nUTnvCvzFJRmdvuXzUzp5FgBPP/00sVgso9MfahTCIiKDRCwW46GHHsJ1XYonTSNQdVpW5lM0diL+\n8ioOHTrE8uXLszKPoUIhLCIySDz//PPs378ff2kFpZPPz9p8LMtH6ZTZAPzhD38gEolkbV6DnUJY\nRGQQqK2t5ZlnngGg7PwFWP7snnQhNOpsAlWn0djYyNKlS7M6r8FMISwiUuASiQQPPfQQsViM8Fk2\noRGjsj5Py7IonTIH8NbAOzo6sj7PwUghLCJS4JYvX86OHTuwisKUTpubs/kGa0YTPO0MWltbefHF\nF3M238FEISwiUsAaGxt5/IknACibPh9fqPezImWDZVmUnOetDb/44ou0tLTkbN6DhUJYRKSAPfro\no3S0txOqGUPR6PE5n39o+BkEa0bT2dnJ888/n/P5FzqFsIhIgVq/fj3vvvsulj9A2fkXndJhKU9F\nat/w0qVLaWhoMFJDoVIIi4gUoI6ODh555BEASs6bg7+k3FgtwerhhEadTTQa5Y9//KOxOgqRQlhE\npAD99re/5ejRowSqT6d4whTT5VBy7kwAVqxYoXHD/aAQFhEpMHv37vXG5loW5ecvwLLM/5QHq4YT\nqDqNtrY21q1bZ7qcgmH+kxMRkT5zXZdHH33UOzTl+ClZOzTlQITPPgeAN954w2whBUQhLCJSQFau\nXInjOFhFYUqSJ1LIF0WjJ4Dfz7Zt26irqzNdTkFQCIuIFIi2tjYef/xxAMqmXogvVGS4ouP5QkUU\nnekNk3rzzTcNV1MYFMIiIgXi6aefprm5mcBpNRSNnWi6nF6Fz7YBeOutt4jH44aryX8KYRGRAvDR\nRx8d64w1w9yY4JMJnnYG/rIKGhoa2LRpk+ly8p5CWEQkz7muyyOPPILruoTHTc6rzlg9WZbV3UFL\n5xo+OYWwiEieS++MlTqPbz4Lj50ElsWGDRtobGw0XU5eUwiLiOSx9vZ2nkidoGHq3LzrjNUbX7iE\n0BljSSQSrFixwnQ5eU0hLCKSx55++mmampoIDBtB0dhJpsvps/RN0q7rGq4mfymERUTy1L59+7o7\nY5XNzN/OWL0J1YzGFy7h4MGD7Nixw3Q5eUshLCKShxKJBI888giJRILwuMkEq4abLqlfLJ+P8Fne\ncCUdQevEFMIiInlo+fLl7Nixo2A6Y/UmNWZ4zZo1tLW1Ga4mPymERUTyTENDA4+nOmNNn18QnbF6\n4y+tIHj6KKLRKG+//bbpcvKSQlhEJM88+uijdLS3EzpjDEWjx5su55RozPCnUwiLiOSRd999l7Vr\n12IFgpTNvLigOmP1pmjUWVihIvbs2cOePXtMl5N3FMIiInmira2NRx99FIDS8+bgLykzXNGps/wB\nwmO841yrg9YnKYRFRPLEk08+SWNjI4FhIwiPn2y6nIxJbZJ+++23iUQihqvJLwphEZE8sH37dl5/\n/XXw+SiftRDLGjw/z4HKYQSqhtPe3s66detMl5NXBs+nLCJSoCKRCA899BAAJfYMAhXVhivKvNSY\n4bfeestwJflFISwiYtgzzzxDXV0d/vIqSs6ZabqcrCgaMx58PrZs2cLRo0dNl5M3FMIiIgbt3buX\n5//4RwBvM7Tfb7ii7PCFwhSNPAvXdXVShzQKYRERQxKJBL/5zW9IxOOEx08heFqN6ZKyqihtk7RO\n6uBRCIuIGPLMM8/wwQcf4CsupfS8OabLybrQiDO7T+qwc+dO0+XkBYWwiIgBjuPw+9//HoDy2Zfg\nC4YMV5R9ls9H0VhvzLA6aHkUwiIiOdbW1sYvfvELXNel2J5OaMSZpkvKmXDynMirV6+mq6vLcDXm\nKYRFRHLIdV3+9V//lSNHjhCoGl6wZ0gaqEBFNYHq0+ns7GTt2rWmyzFOISwikkNvvvkm77zzDlYg\nSMXcy7B8g7M39KdJjRl+8803DVdinkJYRCRHamtru48NXTbjIvxllYYrMqNo9Hjw+dm2bRuHDx82\nXY5RCmERkRyIRqP88z//M5FIhKIxE7o7KA1FvlARRaPOBtRBSyEsIpIDTz31FHv37sVXUj4oTlF4\nqsJneR203nrrLRKJhOFqzFEIi4hk2aZNm3jxxRfBsqiYe9mQGI50MsERo/AVl3Lo0CF27Nhhuhxj\nFMIiIlnU1NTEgw8+CEDJ5NkEh40wXFF+sCxf93ClobxJWiEsIpIlsViM+++/n6amJoLDR1JyznTT\nJeWVouQm6TVr1tDZ2Wm4GjMUwiIiWeC6Lg8//DA7duzAFy6h/ILLBtU5gjMhUFZJ4LQaurq6WLNm\njelyjNA3QkQkC1555RWWL1+O5fdTMf+z+ItLTJeUl4b6mGGFsIhIhm3evJnHHnsMgLLZlxCsHm64\novxVdOY48AfYsWMHtbW1psvJOYWwiEgG1dbWcv/9D+C6LiXnnk949ATTJeU1XzBEeIz3Hr388suG\nq8k9hbCISIa0tbXxs5/9jI6OdkKjzqJk8izTJRWE4olTAa+XdEtLi+FqckshLCKSAfF4nAceeIC6\nujr8lcOomLNoyB+Qo68CFdUER4wmEonw+uuvmy4npxTCIiIZ8B//8R9s2bIFqyhM5fwrsAJB0yUV\nlJJJ3trwq6++SiwWM1xN7iiERURO0bJly3jllVfA56Ny3mfwl5SbLqngBEecib+8ioaGBt555x3T\n5eSMQlhE5BS8/fbbPPzwwwCUz1xA8LQzzBZUoCzL6t43/NJLL+G6ruGKckMhLCIyQOvWreNXv/qV\n1xN6yhzCZ9umSypo4bETsUJh9uzZM2SOJ60QFhEZgM2bN/PAAw+QSCQotmdQeu5M0yUVPMsfoHj8\nZMBbGx4KFMIiIv20Y8cO7rvvPmKxGMUTzqP0vDmmSxo0isdPBp+P9evXU1dXZ7qcrFMIi4j0w+7d\nu7n33nuJRCKEz7IpnT5PQ5EyyBcuoWj0BFzX9Tq7DXIKYRGRPvroo4/46U9/SmdnJ0WjJ1A2a4EC\nOAtKkh20li9fTltbm+FqskshLCLSB7W1tfz4xz+mra2N0MizKJ9zqc6KlCWBqtMInj6Krq4u3njj\nDdPlZJW+QSIiJ7F3717uvvtumpubCY44k4q5l2P59POZTanhSkuXLiUejxuuJnv0LRIR+RRbtmzh\nH//xH2lsbCR4+kgq512B5febLmvQC50xBn9ZJUeOHGHt2rWmy8kahbCIyAmsWLGCn957b3If8Hgq\nL1qCFQiYLmtI8A7ecR4AL774ouFqskchLCLSg+u6PPvss/z6178mEY9TPGka5RdcpjXgHAuPnYQV\nLGL37t2D9uAdCmERkTTxeJyHH36Yp556CoCyGfMpm3ahekEbYAWCFE+YAsC//du/Dcp9wwphEZGk\nzs5O7rvvPpYtWwZ+PxUXLqZ4wnmmyxrSSuzp+ErK2Lt3L6+++qrpcjJOISwiAhw+fJh77rmHjRs3\nYoWKqFrwOYrOHGe6rCHPCgQpmz4fgN/+9rc0NDQYriizFMIiMuStXr2a//n97/PBBx/gKymj6tL/\nprMh5ZGiUWcRGjmWzs5OHnvsMdPlZJS6+YnIkNXR0cGjjz7KihUrAG9YTPnsS/AVFRuuTHoqmzGf\nhvoDvPPOO1x66aVMmzbNdEkZkbGeBu5QOfmjiAwKu3fv5he/+AX19fXg81M27ULC4yerA1Yea9/x\nHm1b36Wmpoa7776bUChktB4rA18WhbCIDCmJRILnnnuOp59+mkQigb9yGBUXXEagotp0aXISbiJO\nw2u/J97SyJ/+6Z9yzTXXGK1HISwi0g8HDx7kX/7lX7rHnBZPnErpeXOw/NozVygih2tpevOPBAIB\n7rnnHs44w9y+e4WwiEgftLS08Pvf/57Xli0jEY/jKyqmfPYlhM4YY7o0GYDmtcvp+mgnU6dO5a67\n7jK2CyETIazFPxEZtCKRCK+88grPPvssHR0dAN45gM+7AF9Yna8KVdm0uURqP2LLli288847zJs3\nz3RJA6Y1YREZdBKJBKtWreKpp57i6NGjAARHjKZs2gUEKk8zXJ1kQseH79O6YQVVVVX85Cc/obg4\n9wtV2hwtIpImkUiwceNGnn76afbu3QuAv3IYZVPnEqoZbbg6ySTXdWlc/hyxo/VceOGFfOtb38KX\n49NLKoRFRIDGxkbeeOMNXn/99e41X1+4hNLz5lA0diKWpeMSDUaxpqM0Ln8ONxbl4osv5pvf/GZO\ng1ghLCJDluu6bN++nddee41169Z1H9zfV1pB8fhzKR43RacdHAKihw/StPIl3HiMBQsW8I1vfCNn\nQawQFpEhxXVdDh48yPr163nzzTc5cOCA9wfLIjRyLMXjJhMccaYOuDHERA7X0rzyZdx4jIULF/L1\nr389J0GsEBaRQS8Wi7Fjxw42bNjAxo0bqaur6/6bL1xC+OxzCJ99Lv6SUoNVimmRQ7U0r/KC+JJL\nLuGmm27KehArhEVk0HFdl7q6OhzH4b333mPz5s3dw4sArFARoZoxFJ15NqEzxmLluDOO5K/IoQPJ\nII6zaNEibrzxxqwGsUJYRApee3s7u3fvZvfu3ezatYtdu3bR1tZ23HP8FdWEzhhD0cixBIaNUEcr\nOaFI/QGaVr0MiewHsUJYRApGLBajvr6e/fv3s3//fj7++GP27dtHbW0tPX8+rKJigsNGEBoxitAZ\nY/GXlhuqWgpRpH4/TategUScq666iuuuuy4r81EIi0heSSQSHD16lEOHDlFfX8+hQ4c4ePAg+/fv\np7a2trsH83F8PgJVwwkOG0Fg2OkEh43AV1ymzlVySroO7KV59VKGDRvGz3/+86zMQ4etFJGccV2X\n1tZWjh49SkNDw3GXw4cPU19fz5EjR3oP2iRfSRmB8mr8FVUEKqrxl1cTqByG5ffnsCUyFASqhwN8\nYitLvlEIiwxRkUiE9vb27ktbWxstLS00NTXR0tJCS0sLzc3N3Zempiai0ehJp+sLl+ArKcNfWoG/\ntNy7rqgiUF6FFQjmoGUihUMhLJJHXNclHo8TjUaJxWLHXfe8HYlEPnE7EokQiUTo7Oykq6ur+zp1\nu7Ozszt0+xKoPVnBkBeyxaX4wiX4i0vxFZfgKy5LBm65Tgso0g8Z+2/Zs2cPlmUdtx+nt83lrut2\nX8Dbh5S6Tr/E4/HuH6TU/dQlFosddz/1WOpHKnW/L5feppO67bouiUTiuJpT91Pt8/l83e1O3ff5\nfPj9fgKBQK/XwWCQYDBIIBDovp1+/9Ou/X4/fr8fn89HIBDonlfPOlL1pX8GPduS3qbU+9zbJb3d\nPR/7NKmaUu9J+qXn+9Lzdm+XYDCI3+/P2b7C9O9RKtzSwy499NIvXV1dn3ist9ecKGBztvnM8mGF\nQviCRVjBkBewRWF8RcX4isJYoWO3fUVhfOESrclmgOu6tG9fT+dHu0i0t5gup198JeWEx06kZPIs\n7bPPkIyF8A9+8INMTapgpBYSJLdSCyLpl1SI91wYSV846rkwkb5QkUgkiEajx62FGtuXZFng82P5\n/Vg+f/K2z7tOXvCn/d0f8G77A95zAgEsfwArEEzeTl4Hgt4l6AUvOVygkWPat6+n/f0NxuYfDoeZ\nNWsWkydPZsWKFezYsaPPr020t3TXXjpldrZKHFIyFsKBytNw3U9fM8J1vR+Y1JoaVrJ/tnXsccvy\nxgBalreknnrc5/MG5VvHrrsf8/m81/j83n2/P3k/9ffUj1n6/dS1dWyaPaefWpO0rLQak/V6DQI3\n2S5cSK3huy4kEriJBLgJ3ET82P1E/Nj9eOp2HDd+7PrY8+Nw3P3U9LxrEq73nqdu4x57n9OvU1Lv\n73HvO59sa/r7nrzvvfz4zwksTvQb7s06+V646e9Nwrud3pbkpbudbvJ9iqfer0T3+0RywSdnCz++\nE4Rd8trqvk5/rOf9tIBM3ic5zWOPp+bhG/RjYJtWvkykbp/pMoas8ePHc8sttwAwb948vv/971Nf\nX9+vabS/v4H29zcQqhlD5cVXZqPMISNjIewLFx/74T1Oz/upQOs9hNOD4Nh9H/is7mBND83uMO0O\n3dQPWY/Q9fvTXuc//vU9Q7jPawep+o9vXXfL3URa+KbCMz1wE58Swl740PN+WoCR2jSevN39bncv\nFPQsNxXCpC1Y0GPBxzr+fT8ucI8P4OMXSHp5H9JD13Wx0kLZTQ/enu+Te2wh5bhwTi3AxOO48dgn\nFzIyxbKwAiHvO5P8rvQM1e5A9qXdD/QM5fT7fvD1uO8PpH1/tUYqueFP64me2qUl5mQshCN1H2dq\nUnnASgZy2j7VT4QPqcQ7trYHaWvC2kydS+n7ynvbHJ36HHvbHO267vGbn10XN9qF2/9+SwPXvQDZ\nY5PzcVtyUmvMvrSFgPQ19ORzAsGTbI4OGl3bHuprTm3b1hndHP3BBx/w5JNPMmPGDJ5//nkOHjzY\n72mUnHu+NkdnSMYWvzds2OD2tWMWHN8hK3Wd3hGrZyetRCJxXKer3jpU9exw1bOz1olu97xkimVZ\nvXbO6tkpq+ftTHTMSu+clf7en6ijWfr73HNf6Ynuf9o+U9d1e+2Qld4x62Sd107UMSv9fUhdMrEm\nmfqOpb436d+hnp2wTtQpq7fOWb31ZO7ZMSvXvDAuSnbMCiU7ZhUd64QVKsZK3U511FKnrIzIl45Z\nlmX1u99DIXXMine0cfTFx6muruaf/umfsjKPvDpYx8yZMzM1KaN6C6TeAqtnp5+e97WZp/D4fD5C\noRChUCin802tiadCumdv6RMNUTrRMKXUkKTehih1dHTQ0dGBG4vixqLQAX1d7LQCQW84Urj0hEOU\nfKGirL5Xg4FlWZROma01SQE0TvgTLMvqXrsSyQXLsrq3hORCIpGgo6PjuIN0tLe309raSmtra/fB\nOXoesCMajRJvaSLe0nTitgRD+EvK8SXHDHcfrKO8Gl9xSd6vPYnkmkJYZIjx+XyUlpZSWtr38++6\nrkt7e3uvh6w8evQoR44cob6+nq6uLmJNR6DpyCemYQWC+CuqCZRXedcV1fgrqr3xxwpnGaIUwiJy\nUpZldQf3mDFjen2O67q0tLR0n7ghdRKH1AkcWltbiR2tJ3b0+OEwvnBJ94kbAsNGEKwarv3Pcsri\nrc2mS+gThbCIZIRlWVRUVFBRUcHEiROP+5vrujQ3N3efxjB12bdvH21tbUQO7CVyYG9qQgQqhhFI\nnsowOOJMfMHc7qeXwhZtOEzz6qUAnH/++Yar+XQ6laGIGOO6LnV1dezatav7sm/fvuMPiWr5CJ5+\nBqEzxlJ0xlj8ZRXmCpa8F208TNNbL+JGu5gzZw7f/va3CQSys76p8wmLyKDT2dnJnj17cByH9957\nj507dx43lMZfXkXojDEUnTmOQPXp2p8s3WKNR2hc8QJupItZs2Zx6623Zi2AQSEsIkNAS0sLmzZt\nYsOGDWzevJn29vbuv/krh1E8bjLhsRO1H3mISw/g888/n+9+97tZDWBQCIvIEBOLxdi5cyfr1q1j\n1apVtLR4B7uwAkGKxkykePxkApXDDFcpuRZrOkLjW14Az5w5k+9+97s5GfKnEBaRISsajfLuu+/y\n2muv4ThO9+OB02ooHj+FotHjBv3JMARiTUeTAdzJjBkzuO2223I25l4hLCIC7Nu3j2XLlrFixQo6\nOzsBb1N12dS5hGpGG65OsiXe1kzDG8/idnUyffp0brvttpwe8U4hLCKSprOzk1WrVvHss89y5Ih3\nwJDgiDMpmzqXQNVphquTTHJdl6aVLxGt38/UqVO54447cn7IWYWwiEgvIpEIS5cu5Zlnn6Uj2ZGr\naOwkSqfMwV/S9yOFSf7q/PgDWtYso7S0lB//+MdUVlbmvAaFsIjIp2hpaeGZZ57h1Vdf9c6Q5vdT\nMnEqJefMVG/qApaIRmhY+lsSne3ceOONXH755UbqUAiLiPRBXV0d//mf/8maNWsA8JdVUn7BZQSr\nhxuuTAai9b236di9lQkTJvC3f/u3xs5YpxAWEemHnTt38pvf/IaPP/4YfD5Kp8ymeNJ0HfCjgEQb\nD9P4+jNYwN///d9z9tlnG6tFISwi0k+RSIQnnniCpUu9YwsHTx9F+ZxL8RdrX3G+c12XxjeeJdZw\niCuvvJKvfOUrRutRCIuIDNDGjRt58MEHaWlpwQoVUT5rIUWjzjZdlnyKjg+207pxJVXV1fzkxz+m\nuLjYaD2ZCGGNZBeRIWnmzJncfffdTJ8+HTfSRfPqV2lZvwI3FjVdmvQi0dlB29Z3AfjKl79sPIAz\nRSEsIkNWVVUVd955J1/+8pcJBAJ07nmfhuXPEe9oM12a9NC6ZQ1uNMK0adOYO3eu6XIyRiEsIkOa\nz+djyZIl/OhHP6KmpoZ401Fvv2Nzg+nSJClyqJauj3YSCAS44YYbBlVHOoWwiAgwduxYfvjDHzJx\n4kQSHW00Ln+OyOFa02UNeW4iTuvGlQB8/vOfp6amxnBFmaUQFhFJKi8v56//+q+ZPXs2bjRC04oX\n6fz4A9NlDWkdu7YSb2mkpqaGq666ynQ5GacQFhFJU1RUxHe/+10+85nPQCJBy5pltO/cbLqsIcmN\nx+lIvvdf+cpXcnZ2pFxSCIuI9ODz+fjqV7/KddddB0Db5ndo3bQajcTMra6PPyDR1cGYMWOYMWOG\n6XKyQiEsItILy7K46qqruOWWW/D7/XTs2kLLmmW4iYTp0oYE13Xp2LUFgCVLlgyqzljpFMIiIp/i\noosu4q677qK4uISu/R/Ssm45rqsgzrbo4VpiTUeoqKhg3rx5psvJGoWwiMhJTJkyhb/6q7soCofp\n2reb1g0rtWk6y1JrwYsXL875eYJzSSEsItIHEyZM4Ht33kkoFKJzzw7atI84a2KtTURqPyIQCLB4\n8WLT5WSVQlhEpI/OPfdcbrvtNgKBAB27t9K+ba3pkgaljl1bAW9XQGVlpeFqskshLCLSD9OnT+fb\n3/42Pp+P9h3v0fb+RtMlDSqJSBddex3A65A12CmERUT6ac6cOdx8881YlkX7trW0J/dfyqnr3PM+\nbjzG1KlTGTNmjOlysk4hLCIyAPPnz+drX/saAG2bVtPx4fuGKyp8biJBx+5twNBYCwaFsIjIgC1a\ntKj7xPKtG1bQdWCv4YoKW9f+D0l0tDFq1CimTZtmupycUAiLiJyCK6+8kmuuuQaAlrVvEGs6arii\nwpR+cI4rr7wSn29oxNPQaKWISBZdffXVzJ8/HzcWpentV0h0dZguqeDEjtYRazhEWVkZF198sely\nckYhLCJyiizL4utf/zrjx48n0d5K8zuv4SbipssqKO07vbXgyy+/nKKiIsPV5I5CWEQkA0KhELff\nfjtV1dVEDx+kdeMqHcyjj+JtLURq9+L3+7niiitMl5NTCmERkQyprq7mjttvJxgM0rlnB50fbDNd\nUkHo2L0VXJd58+ZRVVVlupycUgiLiGTQ+PHj+cY3vgFA66bVROo+NlxRfnPjMTqH0ME5elIIi4hk\n2Pz58/n85z8PrkvzmmXEWppMl5S3ug7sxY1GGDduHGeffbbpcnJOISwikgVf+MIXmD17Nm40QvPb\nr5CIdJkuKS+l1oIvueQSw5WYoRAWEckCn8/HzTffzJgxY4i3NtGy/k111Ooh3t5KtH4/gUCA+fPn\nmy7HCIWwiEiWhMNhbr/9doqLi4kc2EunDm15nM6PdgIwe/ZsSktLDVdjhkJYRCSLRowYwY033ghA\n2+bVxJobDFeUH1zXpWuvF8JDdVM0KIRFRLJu/vz5LFy4EDcep3nNMtx4zHRJxsWO1BFva6a6upqp\nU6eaLsc4nUBWAAAQiElEQVQYhbCISA589atfpaamhnhzA21b1pgux7hUh6yLL754yBwnujdDt+Ui\nIjkUDof51re+hd/vp2P3NrpqPzJdkjFuLErX/g+Bob0pGhTCIiI5M378eK699loAWta9Sbyj3XBF\nZnQd2IMbizJx4kRGjhxpuhyjFMIiIjn0uc99jqlTp+JGOmlZ98aQHLaU2hS9cOFCw5WYpxAWEckh\nn8/HX/7lX1JeXk60/gAdOzeZLimn4m0tRA/VEgwGmTdvnulyjFMIi4jkWFVVFd/85jcBaNu6jujR\nQ4Yryp3U2OA5c+ZQUlJiuBrzFMIiIgbMnDmTz372s+AmaFn7xpAYtuS67pA/TGVPCmEREUOuu+46\nRo0aRby1ifb3N5ouJ+uih2tJtLcybNgwpkyZYrqcvKAQFhExJBQKcdNNNwHQ7mwi1nTUcEXZ1Zk8\nQtbChQuH9NjgdHoXREQMsm2bxYsXe5ul17+F6yZMl5QViWike2ywekUfoxAWETHsuuuuo7q6mljD\nITp2bzNdTlZ07f8Q4jHOOeccampqTJeTNxTCIiKGFRcX8xd/8RcAtG9dS7ytxWxBWdCVtilajlEI\ni4jkgVmzZnHhhRfixmO0bFg5qA7iEWttInrkIKFQiLlz55ouJ68ohEVE8sT1119PaWkp0fqP6dq3\n23Q5GdO5xxuWNHfuXIqLiw1Xk18UwiIieaKyspI///M/B6B109skujoMV3Tq3ESCruTY4Msuu8xw\nNflHISwikkcWLlzIlClTcCNdtG5abbqcUxY5uI9EVwejRo1i0qRJpsvJOwphEZE8YlkWX/va1wgG\ng3Tt203XwX2mSzolnXt2AN4RsizLMlxN/lEIi4jkmZqaGr7whS8A0LphJW4sariigYl3tBGp24ff\n72fBggWmy8lLCmERkTy0ZMkSzjrrLBIdrbTvKMxDWnbu3Qmuy6xZs6isrDRdTl5SCIuI5CG/388N\nN9wAQPvOzcRamgxX1D/pJ2tYtGiR2WLymEJYRCRPTZo0yTu4RSJB63tvF9TY4ejhWhJtzQwbNoyp\nU6eaLidvKYRFRPLYddddR3FJCdH6j4kc2Gu6nD5L75ClkzWcmN4ZEZE8VllZybV/9mcAtG5ajRvL\n//MOJyJddO3fg2VZOm/wSSiERUTy3OWXX87YsWMLppNW175dkIgzdepUTj/9dNPl5DWFsIhInju+\nk9YmYq3520nLdV06kpuiL730UsPV5D+FsIhIAbBt2xtrm0jQlsedtGKNh4k3HaWsrIxZs2aZLifv\nKYRFRArEF7/4RYpLSojUfUyk9iPT5fQq1SFrwYIFBINBw9XkP4WwiEiBqKys5M9SR9La9HbeddJy\nY9Husz9pU3TfKIRFRArI4sWLGTNmDIn2Vtqd90yXc5yu/R/ixqJMnDiR0aNHmy6nICiERUQKyHGd\ntJxNxPOok5Y6ZPWfQlhEpMCcc845yU5acVo2rMyLTlqxlkZiR+oIh8PMmzfPdDkFQyEsIlKAvvSl\nL1FWVkb00AG6Ptppuhzad3ibxufNm0c4HDZcTeFQCIuIFKCKigq+/OUvA9C6+R0SnR3Gaom1NNK1\nbxc+v5+rrrrKWB2FSCEsIlKgLr74YqZOnYob6aJ182pjdbRvWweuy6JLL6WmpsZYHYVIISwiUqAs\ny+LGG28kFArRtW83XQf35byGaONhuvZ/SCAQ4E/+5E9yPv9CpxAWESlgI0aM4JprrgGgdeNK3Fg0\np/Nv37oOgCuuuIJhw4bldN6DgUJYRKTALVmyhLPOOotEeytt29blbL7RIweJ1O0jHA5rX/AAKYRF\nRAqc3+/npptuwrIsOnZvJdpwKOvzdF2Xtq1rAW8hoKKiIuvzHIwUwiIig8C4ceNYsmQJuC6t61fg\nJhJZnV+0fj/RwwcpLS3lc5/7XFbnNZgphEVEBolrrrmG4cOHE2s6QseuzVmbT/pa8FVXXUVJSUnW\n5jXYKYRFRAaJcDjMjTfeCEDbtvXEW5uzMp/Igb3EGg9TWVnJFVdckZV5DBUKYRGRQWT69OlcdNFF\nkIjTtHopiUhXRqfvugnatnlrwVdffTVFRUUZnf5QoxAWERlkrr/+ekaOHEm8uYHm1Utx45k75WHX\nvt3EWxoZPnw4ixYtyth0hyqFsIjIIFNWVsZdd91FVXU10cMHaVm7PCMneXATcdq2rwe8/c+BQOCU\npznUKYRFRAah4cOH89+/9z2Ki4vp2v8hbZtWn3IQd+zaQqKthVGjRnHxxRdnqNKhTSEsIjJIjR07\nljvuuINAIEDH7q107Nw0oOm4rkvbtnW0bXkXgGuvvRafT/GRCXoXRUQGscmTJ3PzzTdjWRZtW96l\ns5+nPXTjcVrWvkH7+xuwLIsbbriBOXPmZKnaoUcb9EVEBrkLL7yQxsZG/v3f/52WdW/hKyomVDP6\npK9LRDppXv0q0cMHKQqHufU732HGjBk5qHjosDI1ITcTe/1FRCRrHn/8cV544QUsf4CKCxcTHHEm\n1gk2K8dbm2ha9Qrx1iaqq6u58847Oeuss3JccX6zLOuUM1QhLCIyRCQSCR588EFWrlwJgBUIEjit\nhtDpIwkOH0Wg+jQsy0f0SB1Nby/FjXQyduxY7rzzTp0hqRcKYRER6ZdYLMbvfvc71q1bR21t7XF/\ns4IhgsNGEDlUC4k406dP5zvf+Q7FxcWGqs1vCmERERmwhoYGtm3bxvbt29m+fTv19fXdf1u8eDHX\nX389fr/fYIX5TSEsIiIZc/jwYbZv304oFGLu3LlkIGMGNYWwiIiIIZkIYY0TFhERMUQhLCIiYohC\nWERExBCFsIiIiCEKYREREUMUwiIiIoYohEVERAxRCIuIiBiiEBYRETFEISwiImKIQlhERMQQhbCI\niIghCmERERFDFMIiIiKGKIRFREQMUQiLiIgYohAWERExRCEsIiJiiEJYRETEEIWwiIiIIQphERER\nQxTCIiIihiiERURERERERERERERERERERERERERERERERERERERERERERERERERERERERETyh9WX\nJ9m2fQdwKzAK+AD4e8dxnujja+8Hvg0schznzYEWalJ/2m/b9t8Bf9vLn1odx6nIWpFZ1N/P37bt\nSuCnwBeAILAS+JbjOB/moNyM6+fn/wZwyQkm9SPHcX6UjRqzZQCf/WLg74CpeMemXwb8d8dxdmW/\n2swbQPuvAP4RmA40Ag8Df+M4Tjz71WaWbdt+vM/y+/Thu2vb9hzgZ8AFQBvwFHCn4zgdWS41K/rb\n/uRrLgGeALocxxnXl/mc9AQOtm1/C7g7Wcw04NfAv9u2/dk+vPYC4BuA25di8tEA278POKPHZXx2\nK82OAbb/D8BE4DJgIVAOPG/bdp8W+vLJANr/p3zys78QiOIFUsHob9tt254NvASsAeYCi4AK4FXb\ntktzUHJGDaD9c4EXgG3ALODrwC3APbmoN5Ns264BlgLXADFO8htu2/ZI4FW8BZULgOuAK4D/l91K\ns6O/7U++5n/gff7NfXl+SuAkE7WA/wn80nGcR5MP/9y27UuTj7/yKa/1431pHwa+2deC8skptD/h\nOE59LmrMpoG037btK/F+gMc6jnMk+dhX8H6UQkBXLmrPhIG033Gchl6mcz/wvOM4b2Wz3kwa4Hf/\ni0Cj4zh3pk3nDmAjsAB4ObtVZ84A238rcAC4yXGcBLDdtu3vA/fatv0PjuO05KL2DPkScAT4PHCo\nD8+/FegEvuE4TgzYatv2ncAfbNv+QQFuBetX+23brsJb4FoE3AQs6euMTrYmfC7eZpieX7hXgQW2\nbRd9ymtvBYqB/9PXYvLQqbR/MBhI+z8PLEsFMIDjOHscx3nacZyCCeCkU/78bdteCFwN3JX58rJq\nIG1P8Mk1gEjyutC2hg2k/VOA1ckATnkeKMJbCCkkTzmOc63jOK19fP5i4I1kAKe8hve5L854ddnX\n3/a3AbMcx1lLH3fzppwshCcmr/f0ePyD5Gt73cRq2/Zo4Ed4SwaR3p5TIAbU/kFkIO2fCuy0bfuv\nbNt2bNuut237cdu2h2exzmzJxOf/A+BJx3F2Z7CuXBhI2x8GSmzb/p5t22Hbtkvw2u9QYJviGVj7\nfXibLtOl1qImZKyyHHAcZ38/XzKBHu+V4zhtQD0wKUNl5Ux/2+84TtRxnMMDmdfJQrg8ed3W4/HU\n0sGJOhr9E/AHx3HeGEhReWSg7S+xbfsXaSH0tG3bE0/w3Hw2kPaPAK7FC+MvAl/D2ye6PLmLopAM\n9PMHwLbt84HPAPdmuK5c6HfbHcfZjrdP/G+Sz2sGzgeW9FhDKgQD+ex3ALN7PDajx/QGq3I++V6B\n934VZIfUXDlpx6z+sm37v+H1Dr3zZM8dpFrwvnib8DZDXg+MBVYW6NpgfwWBDuCrjuOsdxznebwO\nKpOBk3bmG2S+A7zlOM4m04Xkgm3bU4HH8daIL8TbDLkXr1PeYA8hgF8Ctm3bf5PcEjABb3dcO17H\nPJFP+NSOWUBT8rrnkkxlj78DkOwB+QBwVy+r5gXXM5Z+th/AcZyf4XXTT9lm2/YW4CPgL4H/leki\ns6jf7cdb+9ntOE76PsBVePuGpgEvZrTC7BpI+wGwbduHt3/8J1moKxcG0va/w/vsb089YNv2euAg\n3haRn2e4xmwayP/+8mSP6nvx3os6vL4x/4++dW4qZE30vsZbyaf8n8jJ14R3Jq977s+YhLdk13M/\n1xxgDPBr27ajtm1H06bxmm3bzqkUa0B/29+r5P6FI3jDVQrJQNq/Ezitx2M+vIWw5oxWl32n8vlf\nhPc+vJCFunJhIG2fDLyf/kCyR3A9x/axFooBffaO4/wKGIa39Ws03hj5arwtY4PZTnp8xrZtVwPD\nge1GKioQnxrCjuM4eB0RPtfjT/8VeNVxnJ6bWN7F2xc4I+3yX5J/uyntdkEYQPuxbfse27Zv7PHY\nWLwvY0EthAyk/XjjROfZtp0exBclrwvqh2iA7U+5FGhxHGdrturLpgG2fR9wTvoDyQO3jEz+rWAM\n8H9/mm3bX3IcJ+I4zoHk1qAvAR85jrMh+1Ub9RJwqW3b4bTH/gtej/mCGZpmwsk2R4PXy/lfbNte\nBbyJ19lmEd5BGLBt+x7gfMdxljiO0443UL2bbdvtyZsfFuhRc/rc/uTzfcADyXGGy/F+gO7FGz/4\nKIWnv+1/DPgr4Le2bX8br6PWL4AVjuOsynHtmdDf9qdMwtsfWsj62/b7geds2/4HvO9BGPgh3prj\nU7ktPSP62/5xwGPJvh/P4Y2XT40SKSi2bc8EqpJ3fcA427YXJe+/jbe5Pb3t/4y36f0h27Z/iLcV\n4H8Dv3Ic52Cu6s6U/rY/eXCPycm/jwLCyTHlFl72nfC34KQdsxzH+Tfg9uRMdwB/DlztOM7q5FP6\ncjSoQhsj2G0A7f8fwD/gjQvdAjyDtz94oeM4BbdvpL/tdxwngtcjuAFYjXf0rJV4axAF5xS+/1UU\n+L6wAXz2f8Q7wtBVeAfoeBMoAxYX4MEaBtL+Z/GC6Da8zfJ/B9zqOM5/5K7qjPm/eMPKluGtrN2Q\nvP0avRwF0HGco3gd8UbhbfF6Inm5ncLUr/bjbTFJPf8qvJWP15P3b8hZ1SIiIiIiIiIiIiIiIiIi\nIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIZ9/8B05w8OE9B3rsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fdb909a8ed0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "plt.figure(figsize=(8, 4.5))\n",
    "sns.violinplot(seq_ident.values, vert=False, color = sns.color_palette()[5])\n",
    "plt.xlim((0.4,1.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for_scoring['ssb_si_score'] = for_scoring.ssb_p_percent_seq_ident / seq_ident_stats_simple_cutoff_10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">O32583</th>\n",
       "      <th>1F0Z</th>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>1.208995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1ZUD</th>\n",
       "      <td>66</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>1.208995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">P00350</th>\n",
       "      <th>2ZYA</th>\n",
       "      <td>468</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "      <td>1.201245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2ZYD</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "      <td>1.201245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3FWN</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.995726</td>\n",
       "      <td>466</td>\n",
       "      <td>466</td>\n",
       "      <td>1.203828</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     u_seq_len  p_resolution  ssb_p_percent_seq_ident  \\\n",
       "u_uniprot_acc u_pdb                                                     \n",
       "O32583        1F0Z          66           NaN                 1.000000   \n",
       "              1ZUD          66          1.98                 1.000000   \n",
       "P00350        2ZYA         468          1.60                 0.993590   \n",
       "              2ZYD         468          1.50                 0.993590   \n",
       "              3FWN         468          1.50                 0.995726   \n",
       "\n",
       "                     ssb_p_aln_coverage  ssb_p_aln_coverage_sim  ssb_si_score  \n",
       "u_uniprot_acc u_pdb                                                            \n",
       "O32583        1F0Z                   66                      66      1.208995  \n",
       "              1ZUD                   66                      66      1.208995  \n",
       "P00350        2ZYA                  465                     465      1.201245  \n",
       "              2ZYD                  465                     465      1.201245  \n",
       "              3FWN                  466                     466      1.203828  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_scoring.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scoring resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    3708.000000\n",
      "mean        2.293873\n",
      "std         0.984204\n",
      "min         0.890000\n",
      "10%         1.640000\n",
      "25%         1.900000\n",
      "50%         2.200000\n",
      "75%         2.600000\n",
      "90%         3.000000\n",
      "max        28.000000\n",
      "Name: p_resolution, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "resolutions = for_scoring.p_resolution[pd.notnull(for_scoring.p_resolution)]\n",
    "resolutions_stats = resolutions.describe(percentiles = [.1,.25,.5,.75,.9])\n",
    "resolutions_simple_cutoff_75 = resolutions_stats['75%']\n",
    "resolutions_simple_cutoff_90 = resolutions_stats['90%']\n",
    "print resolutions_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes.AxesSubplot at 0x7fdb59702790>"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAdQAAAEjCAYAAACLsPDNAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4VPXB/v/3mS3LTFZCFsK+DCCLsgQ3ELeqVatWf+70\n0WIf7derz9Nvl0frAm2h1fq0dvtZq7bghlZra6u4VKVWcQNFimF1wg5hCVuSyayZOef7x8QICkJg\nkjOZ3K/ryjXJmRlyhyuZe87nc87ngIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiEjWMTryYMuyrM4K\nIiIikmkMwzjinnR0ZhAREZGeQoUqIiKSBipUERGRNFChioiIpIEKVUREJA1UqCIiImmgQhUREUkD\nFaqIiEgaqFBFRETSQIUqIiKSBipUERGRNFChioiIpIEKVUREJA1UqCIiImmgQhUREUkDFaqIiEga\nqFBFRETSQIUqIiKSBipUERGRNFChioiIpIEKVUREJA1UqCIiImngsjuASLo1NDSwfv16ysvLqays\nJD8/3+5IItIDGB15sGVZVmcFEUmHSCTCD37wA/bu3du+raioiMrKSiorK6mqqmr/vLy8HLfbbWNa\nEcl0hmEccU+qUCWrPP7447z66qs48rwYnhySLU2QTB70sYZh0Lt3788VbVVVFSUlJTgcmhER6elU\nqNIjrVu3jh//+MdYQMkZl+Aq7oVlWZiREMmWJpLBJpItTSRaUrdmqAU4+K+0x+OhoqKivWD3L1yf\nz9elP5eI2EeFKj1OIpFg5syZbNmyhTz/WHyjJx32OVYySTLUnCrb9sJtJtHSiBWLHvJ5Pp+Pqqoq\nBg4cyFe/+lUKCgrS+aOISAZRoUqP88ILL/D000/jyC+g9OzLMFzHdrydGY+RbNmvbNsKN9HSBMlE\n++Oqq6u55ZZbKC0tPdYfQUQykApVepSGhgZuu+024vE4Raeeh6eib6d9L8uyMKNhksFGWmoXkWze\nR2lpKbfeeit9+vTptO8rIvboSKHqqAvp1izL4pFHHiEej5PTb0inlimkDmRy5nnxlFdTfNoFuErL\n2bt3L7Nnz2b9+vWd+r1FJLOpUKVbe/fdd1m+fDmGJwffmJO69Hs7PLkUTz4fT0U/WlpauOuuu1i+\nfHmXZhCRzKFClW4rGAzyxBNPAOAbfSKO3Lwuz2C4XBSe/CVy+g0lFotx7733snjx4i7PISL2U6FK\nt/XUU08RDAZxl1WRM2CYbTkMh4OCiVPJGzqaZDLJ7373OxYsWGBbHhGxhwpVuqVVq1axcOFCcDjx\njZtMB44b6BSGYeAdcyLeUTVYlsWjjz7Ks88+i47jE+k5VKjS7cTjcebOnQtA/vATcBUU2ZwoxTAM\n8ocfj2/cZDAM/va3v/Hoo49imqbd0USkC6hQpduZP38+O3fuxFlQTP7wsXbH+Zy8QSMonHQWOJz8\n85//5P777yeRSBz+iSLSralQpVupr69n/vz5ABSMm4zhcNqc6OByqgdSdOq5GC43ixcv5t577yUa\nPfTqSyLS/alQpdswTZO5c+eSTCbJHTQCd1ml3ZG+kKd3H4pOuwAjJ5cVK1Zw9913EwwG7Y4lIp1E\nhSrdxhtvvEEgEMCRk4d3VI3dcY6Iu7iMkqlfwZFfwPr165k9eza7d++2O5aIdAIVqnQLjY2NPPX0\n0wD4jj8ZhyfH5kRHzukronjqV3AWlbJ9+3ZmzZpFfX293bFEJM1UqNItPP7440TCYTyV/fBUD7I7\nToc58/IpnnIB7l6V7Nu3j9mzZ7N27Vq7Y4lIGqlQJeP9+9//5v3338dwuvCdcKrt55weLYcnh6LJ\n5+Gp7E8oFOLuu++mtrbW7lgikiYqVMlo0WiURx55BID8URNx5nfvi3sbTheFJ51NTv9hxONx7v3l\nL3n33XftjiUiaaBClYz2l7/8hb179+IqLiNvyHF2x0kLw+GgYMJp5A0bg5lM8vvf/55XXnnF7lgi\ncoxUqJKx1q9fz6uvvgqGgW/8ZAwje35dDcPAN+ZEvKMnATBv3jz+8pe/aKlCkW4se16hJKskk0nm\nzp2LZVnkDR2Nu7jM7kidIt8/loLxp4Fh8Nxzz/HII49oqUKRbkqFKhnplVdeYdOmTTjyfXhHjrc7\nTqfKHein8MSzweHk9ddf57777qO1tdXuWCLSQSpUyTi7du3ir3/9K0DqqF6X2+ZEnS+nzwCKJp+H\n4fbwwQcf8Itf/IJIJGJ3LBHpABWqZBTLsnjkkUeIx+Pk9B1MTmU/uyN1GU9ZFcVTLsCRk8eqVau4\n6667aGpqsjuWiBwhFapklEWLFlFbW4vh9uAbe5Ldcbqcq7gXxad/BYe3kI0bNzJ79mx27dpldywR\nOQIqVMkYoVCIefPmAeAdPQlHbr7Niezh9BZSMvVCXEW92LlzJ7Nmz2br1q12xxKRw1ChSsZ46qmn\naG5uxl1WSe7A4XbHsZUjN5+i0y7AXVZJ4759/OQnP6Gurs7uWCLyBVSokhHWrFnDG2+8AQ4HvnGT\nu+3ygunkcHsoOvU8PFWppQp/9rOfaalCkQymQhXbtba2MnfuXADyh5+Aq6DY5kSZw3C6KDzx06UK\nf/nLX7Jo0SK7Y4nIQahQxXbz589n+/btOAuKyPcfb3ecjNO+VOHQ0SSTSe6//34WLFhgdywR+QwV\nqtiqvr6e+fPnA6SGep1OmxNlJsMw8I45Ee+oGizL4tFHH+Xvf/+7lioUySAqVLGNaZo8/PDDJBIJ\ncgcOx1NWZXekjGYYBvnDj8c3bjIAf/3rX3niiSe0VKFIhlChim0WLlzIxx9/jCMnr32ReDm8vEEj\nKDzxLHA4eOWVV3jooYdIJBJ2xxLp8VSoYovGxkb+9Kc/AeA9/iQcnhybE3UvOdWDKDrlXAyni3fe\neYff/OY3xGIxu2OJ9GgqVLHFE088QTgcxlPRl5zqwXbH6ZY85dUUTTkfw5PDsmXL+N///V9CoZDd\nsUR6LBWqdLmPPvqIRYsWYThdqcXvdc7pUXOXllN82oU48rwEAgF++tOf0tjYaHcskR5JhSpdKhqN\n8sgjjwCQf9x4nN4CewNlAVdhCcVTv4LTV8SWLVuYPXs2DQ0NdscS6XFUqNKlnn32WXbv3o2rqBd5\nQ0bbHSdrOPN9FE+9EFdxGQ0NDcyaNYvNmzfbHUukR1GhSpdZu3Yt//jHP8Aw8I2fguHQr186OXLy\nKJpyPu7eVTQ1NfHTn/6UQCBgdyyRHkOvaNIlYrEYDz74IJZlkTdsDO6SMrsjZSWH20PRKefi6TOA\ncDjMPffcw7Jly+yOJdIjqFClSzzzzDPs2LEDZ2EJ3pET7I6T1Qyni8JJZ5E7wE88HudXv/417777\nrt2xRLKeClU63erVq3nllVfAcFAwYaqWF+wChsOBb/wU8vxjMZNJfv/73/Paa6/ZHUskq6lQpVNF\nIhEeeughAPKHH6+h3i5kGAa+0ZPaV6F67LHHePbZZ7X+r0gnUaFKp3rqqafaj+rNHzHO7jg9Ur5/\nLL7xU8Aw+Nvf/sZjjz2m9X9FOoEKVTpNbW0tr7/+OjgcFEycqqN6bZQ3cDiFk1Lr/y5YsIAHHnhA\n6/+KpJle4aRThEIh/vjHPwLgHTkBV1GpzYkkp3ogRaeeh+Fy89577/GrX/2KaDRqdyyRrKFClU4x\nb9489u3bh6u0nLxhY+yOI208vftQNOUCDE8utbW13HPPPbS0tNgdSyQrqFAl7ZYuXcrbb78NDicF\nE07TUG+GcZeUUTz1Qhx5PtauXctPf/pT9u7da3cskW5Pr3SSVsFgkLlz5wLgHV2Dq6DY5kRyMK6C\nYoqnXoizoJitW7cyc+ZMraokcoxUqJJWjz76KE1NTbjLKskbMsruOPIFnPk+ik+7EHdZ21KFd93F\nggULdFqNyFFSoUraLFq0iMWLF2M4XakFHHRZtoznyMmlaPKXyRs6GjOZ5NFHH+UPf/gD8Xjc7mgi\n3U6HXvEsvXWVQ2hsbOQHP/gBoVAI3wmnkjd4pN2RpIOiW9bSsvQtrGSSQYMG8d///d+UlWkhDunZ\njA7sGWgPVY6ZZVk8/PDDhEIh3OXV5A4aYXckOQq5/YZSPPUiHPkFbNiwgZkzZ7Jy5Uq7Y4l0GypU\nOWZvv/02S5cuxXB7KBg/RUO93ZiruBclZ16Mu7wvwWCQe+65h5dfflnzqiJHQIUqx2TPnj08Pm8e\nAL6xJ+HM99mcSI6Vw5NL0annkD/8eCzL4sknn+T+++/XIhAih6FClaNmWRZz5swhEg7jqepPTv9h\ndkeSNDEMB95RNRSeeBaGy82iRYuYNWsWO3futDuaSMZSocpR+9e//sXy5csxPDkUjJusod4slFM9\niOLTL8LpK2LLli3MnDmT2tpau2OJZCQVqhyVhoYGnnzySQB8J5yCIzff5kTSWVyFJRSfcTGeqv6E\nw2F+8Ytf8Pzzz2teVeQzVKjSYaZp8oc//IFYLEZO9SBy+w6xO5J0MofbQ+FJXyJ/5AQsy+KZZ57h\nt7/9LZFIxO5oIhlDhSod9tprr7FmzRqMnFx8J5xqdxzpIoZh4B05jsKTz8Fwe1iyZAk//OEP2b59\nu93RRDKCClU6ZPv27Tz99NMAFIybgiMn1+ZE0tVyqvpTfMbFOAtL2L59OzNnzuTDDz+0O5aI7VSo\ncsSSySQPPfQQra2t5PQfRk6fAXZHEpu4fEWUnH4ROdWDiEaj/PrXv+Yvf/kLpmnaHU3ENipUOWIv\nvfQSa9euxZGbj2/sSXbHEZsZLjcFk87EO3oSGAbPPfcc9957L6FQyO5oIrZQocoR2bJlC88++ywA\nBRNOw+HJsTmRZALDMMj3j6Xo1PMwPDnU1tYyc+ZMtmzZYnc0kS6nQpXDSiQSPPjggyQSCXIHjsBT\n0dfuSJJhPOXVlJxxCa6iXjQ0NPCjH/2IxYsX2x1LpEupUOWwnn/+eTZt2oQj34d3zCS740iGcnoL\nKD79K+T0H0o8Hue+++7jT3/6E8lk0u5oIl1ChSpfaMOGDTz33HNA21Cv22NzIslkn1wL13f8yWA4\neOmll/j5z39OMBi0O5pIp1OhyiHF43EefPBBTNMkb8goPL372B1JugHDMMgbMoqiKV/GyMll5cqV\nzJgxg7q6OrujiXQqXWBcDumpp57ixRdfxOkrpOTMSzFcLrsjSTeTjIRoXrSAxL5dAJxyyilcccUV\n9OrVy+ZkIkemIxcYV6HKQdXV1TF79mwsoPi0C3H3qrA7knRTVjJJeM1SwnUrwEzi8Xi44IILOP/8\n88nN1cIgktlUqHJMYrEYd9xxBzt37iTPPxbfaB2IJMcuGQoSWvE+sfoNAJSUlHDllVdy8skn43Bo\n9kkykwpVjsnjjz/Oq6++irOwhJIzLsFwOu2OJFkkvnsHodpFJBp3AzB48GCmTZvGsGG6nq5kHhWq\nHLVVq1Zx9913g+Gg+PSLcJeU2R1JspBlWcQ21xFauQQzGgbg5JNP5oorrqCsTL9zkjlUqHJUIpEI\nt99+O7t37yZ/xDi8x02wO5JkOSvRSvjjjwjXLQczidvt5vzzz+fCCy/U/KpkBBWqdJhlWcyZM4c3\n33wTV1Evis+4GEPzWtJFkuEgoRUfENu6HoDi4mKuuOIKTj31VM2viq1UqNIhO3bsYO7cuaxevRoc\njrYl5ErtjiU9UOuenbTULmo/zWbQoEFce+21DB8+3OZk0lOpUOWIJBIJXn75Zf72t7/R2tqK4cml\nYPxkcvoMtDua9GCWZRHbspbQig/a51dPPPFErrzySnr37m1zOulpVKhyWBs2bGDOnDls2rQJgJz+\nw/CNOVEXDJeMYSVaCQdqCdfVQjKJy+Vqn1/Ny8uzO570ECpUOaRYLMazzz7Lyy+/jGVZOPJ9FIyb\nrCvISMZKhlsIrfyA2JZ1ABQVFXHFFVcwefJkza9Kp1OhykGtWLGCuXPnsmvXLmhbb9V73AQMl9vu\naCKH1bq3gZaP3mufXx0wYADTpk1jxIgRNieTbKZClQO0tLTw5JNP8tZbbwHgLCqlYNwU3KWaj5Lu\nxbIsYlvXpeZXIyEAampquOqqqygvL7c5nWQjFaoAqRefxYsX8/jjj9Pc3AwOJ96R48gbNlanxEi3\nZiUShOtqiQRqsZIJXC4X5513HhdeeCFer9fueJJFVKjCnj17eOSRR1i2bBkA7rJKfOOm4CoosjmZ\nSPokwyFCqz4gtnktAC6XmzFjRlNTU8P48eNVrnLMVKg9mGma/POf/+TpP/+ZWDSK4fbgHT2J3IHD\n6cDvhUi30rq3gdDKJbTu2ta+zel0ctxxxzFx4kQmTJhAUZHeTErHqVB7qPr6eubMmdN+IWdPnwH4\njj8FZ57epUvPYEbDxLZtJFa/kdbd26HtJcswDEaMGEFNTQ0TJkygtFQLl8iRUaH2MIlEgvnz5/P8\n88+TSCRw5OThO+EUcqoH2R1NxDZmLEps+ybi9RuIN2wDy2y/b+jQoUyaNImJEydqsQj5QirUHqSu\nro45c+ZQX18PQO7A4XhHT8LhybE5mUjmMOMx4js2E6vfSHznVjCT7fcNGjSIiRMnUlNTQ1VVlY0p\nJROpUHuASCTCM888w4IFC7AsC6evEN+4yXh697E7mkhGsxKtxHZsSe257tiClUy039evX7/2cu3b\nt6+OOxAVarZbtmwZDz/8MHv37k0t0DBsLN6R4zCcLrujiXQrVjJBfGc9sW0biG/fjNUab7+vsrKS\nmpoaJk2axIABA1SuPZQKNUs1NTXxxBNP8N577wHgKi6jYPwUXMW9bE4m0v1ZZpJ4wzbi9RuIbd+E\nFY+131dWVkZNTQ01NTUMGTJESx72ICrULNPU1MTSpUt5+umnCYVC4HTiPW4ieUNGaYEGkU5gmSat\nu7en5ly3bcSMRdrvy8/Pp0+fPvTt25fq6mr69u1L3759KSoq0l5sFlKhdmOWZbFjxw4CgUD7x44d\nO9rvd5dXUzDuVJzeQhtTivQclmWS2NPQfjqOGWk56OO8Xm97wVZXV7d/rvNfuzcVajeSSCTYtGlT\ne3l+/PHHBIPBAx/kdOEuLSd3wDBy+g3Vu2ARm1iWhRWLkGhuJNG8l2RzI4ngPpLN+w6Yf91fQUHB\nAQX7yW1BQUEXp5ejoULNYJFIhLq6uvYCXbduHfH4gX+IRk4e7l4VuMsqcfeqwFXUS0O7IhnMsizM\naJhk8z4SwcbUbXNb0SZaD/qcwsLCA4aMP/lcyyVmFhVqBtm7dy8ff/wxgUCAuro6Nm/ezGf/G52+\nogMK1OEt1F6oSBawLAszEmor2n2pPdrmfSSDjYcs2uLiYioqKigsLDzgo6Cg4IDPfT6fDo7qAipU\nm5imSX19/QHzn7t37z7wQYYDV0lZqkB7VeLuVY4jJ8+ewCJii1TRtpBoPnBvNhFshP3Oi/0ihmHg\n8/kOWroFBQUUFRUd8LXX61UBHwUVapolk0kikQjhcJhIJPK5z5ubm1m/fj11dXWEw+EDnmu43Klh\n208KtKQ3hkvni6abZVmEVy8lunktZjh4+CdkIUd+Abn9h5I/crxGOLopy7Iww0GSoRbMeAQrFsWM\nRTFjEcxYtO3rCGY8esBpPUfC4XBQUFDQXrJerxeXy/W5D6fTecBtR+7ff9snn3s8HtxuNy6Xq1v+\nXqpQ9xOPxwmHw+0fn5TgFxXkZz//7BznF3Hkedv2PCtwl1XgLCzBMPSusLOFVn1IeM2/u+R7ud1u\nTjjhBMaOHcuiRYtYuXJll3zfI5U/Yhze4ybYHUM6mWWaWPFPy/aT4j2ghOOflvChDprqKoZh4Ha7\n2wv2SG4/+fxQ2z+5dTgcGIbR/vHJnng6thcWFqpQAWpra/n5z3+ehn/JwHC7MVye1K3bg8PlwXB/\n+rWrqBR3r0qc+b7PPbvpnVeI79yShhySCYYNG8aMGTMwDINEIsGdd97Zvpay9Dyein4UnXqu3TEO\nyzKTB+7ltsbBNME0scwkWG23pom1//b2r5NY1sG27/e11fa4/bcnk1jJ5AEXJ+hO5s2bd8Q9mdVj\nj9FoNC3/jiM3b79C9eBwezBc7rZCbduWm4/Dk5uW7yeZzel0tg9dffKOViTTGQ4njpw8LMMAwwCn\nq738DHO/IrT2K9CDFuZBirb9cSZYBy9UK9na4WHq7iar91AhdZRtc3Nz+5BvNBptH9I91DDvJ19H\no9HPHZH7hQwDV/H+BxxV4MjVAUddoSuHfHNzc5kyZQonnXQSr7zyCkuWLME0M+fdt4Z8ew7LMrFi\nMcz4/vOskQPnXfe/L0MKzTCMDg3tHs2Q7/5fH2z7J8O8B9u+/3P69++vId90ME2TWCx22AIOBoOs\nW7fuC0+J+eSgJKdPp8R0BjsOSjIMo2NvuDqZDkrKDlailWS45cByjO8/P/rpnKkV79gonGEYeL1e\nWw5K2r8I9x/lyXQ6KMkmkUiEtWvXtp8ys3bt2oMs2pB7wB6sq7hMizaI9EBWMpE6bSa432kzzfsw\nwwdf2vBQfD7fAafHHOo0msLCQnw+H06ns5N+ouykQs0QiUSCzZs3ty/sEAgEaG5uPvBBTifu0nLc\npamFHVyl5TjcHnsCi0jaWckEyWBT+xKFieZ9JJobMUPNB3280+mkd+/ehyzFzxakS6fhdSoVaoay\nLIuGhoYDCnb79u2feZSBq6g0NURcVoG7tBJnvpYiE8l0lpn8tDibPl0ZKdnSDHz+pdPhcFBZWfm5\nNX4rKipUkhlEhdqNNDc3H7C274YNG0gmkwc8xpHvw92rgtwBfjzl1TYlFRFInf+ZbGn6dG+zbe3e\nZEsTHOQl0jAMKioqPncVmsrKStxutw0/gXSECrUbi8VirF+/vr1g6+rqiEQ+vRZjTv+h+MachCNH\np+iIdJVkSzOxbRuI1W8k0bjnoOdUGoZBeXn55y7hVlVVhcejaZzuSoWaRUzTZOvWrSxZsoQXXniB\n1tZWDE8uvuNPIqfvkG5zpJxId5No3kesfiOxbRtINu094L7evXt/bqi2qqqKnJwcm9JKZ1GhZqkd\nO3Ywd+5cVq9eDYCnoi++cafizNd1FUWOlWVZJJv2EqvfQGzbRpLBxvb78vLyGDduHDU1NYwePZrc\nXI0Q9RQq1CxmWRZvvvkmT/7pT0TCYQyni/xRE8kbcpzWDBbpIMuySOzbRWzbRmL1Gw848tbr9TJh\nwgRqamoYNWqU5jt7KBVqD9DY2Mhjjz3GBx98AICrtJyC8VNwFZbYnEwks1mWSeueBuJte6JmJNR+\nX1FRUXuJjhgxQkfbigq1J1myZAmPPvYYjfv2gcNBvv948oefgKGTt0XaWaZJ6+7txOo3Et+2ETP2\n6YF+paWlTJw4kZqaGvx+v64ZKgdQofYw4XCYp59+mtdffx0AZ0ERBeOn4O5VaXMyEftYZpJ4w7bU\nnuj2TQesY9u7d29qamqoqalh8ODBKlE5JBVqD7VmzRrmzJnDjh07AMgdPBLvqBqtvCQ9hpVMEN+5\nNbUnumPzAdcA7dOnT/ue6IABA3SEvBwRFWoPFo/Hee6553jhxRcxk0kcufn4xp1KTtUAu6OJdJpk\nqJnQqqXEt23ESibat/fr14+amhomTZpEdbUWRZGOU6EKmzdvZs6cOaxfvx6AnOpB+I4/GUduvs3J\nRNLHbI0T/ngZkbUrUhfLBgYPHtw+nFtRUWFzQunuVKgCpBaFePXVV3nmmWeIx+MY7hx8Y04kZ8Aw\nDXdJt2ZZJtGNAUKrlmDFUpcwmzx5Ml/96lcpLy+3OZ1kExWqHGDXrl08/PDDLF++HAB37z4UjJuM\n01doczKRjos3bKNl+aL21Yv8fj/XXnstgwcPtjmZZCMVqnyOZVm8++67zJs3j5aWFgynk/yRE8gb\nOlrXY5VuIdHSRGj5+8S3bwKgrKyMq666ikmTJmnERTqNClUOqbm5mXnz5vHee+8B4CruhW/8FNzF\nZTYnEzk4Mx5LzZOuWwmmSU5ODhdddBHnnXeeFp2XTqdClcP66KOPePjhh9mzZw8YBnnDxuAdOR7D\nqZVhJDNYpkl048eEVn2IFY9iGAZTpkzh8ssvp7i42O540kOoUOWIRKNRnnnmGV577TUsy8LpLaRg\nwmm4y7QghNgr3lBPS+0iks37ABg+fDjTpk1j4MCB9gaTHkeFKh2ydu1a/vjHP1JfX4/hdFFy9mU4\nvbqCjXS9RLCJ0PLFxHdsBlLzpNdccw0TJ07UPKnYQoUqHZZIJLjvvvv48MMPcZdVUjTlAr2ASZcx\n4zHCa/5NZN0qsExyc3O5+OKLOeecczRPKrbqSKFqwkwAcLlcTJ8+nbq6Opp37yCybiX5Q0fbHUuy\nnGWaRDesIbT6Q6x4DMMwmHr66Vx22WWaJ5VuR3uocoAlS5bwm9/8BsPppPjMS3EVFNkdSbJUfOfW\n1Dxp24W8R4wYwbRp0xgwQMtkSubQkK8ckwceeIB33nkHV2k5xVMv1IXLJa0Swca2edItAJSXl3P1\n1VczYcIETTNIxtGQrxyTr33ta6xctYrGvQ1EAsvJH3683ZEkC5jxKOHV/yayfnX7POkll1zCOeec\ng9vttjueyDHTHqocVG1tLT//+c/B4aDkjEtwFZXaHUm6KcuyiK5ffcA86elt86RFRZpSkMymIV9J\nizlz5vDGG2/gKu5F8ekXa4lC6TAzHiO45I324d2RI0dy7bXXap5Uug0N+UpaXHPNNaxYsYLdu3cT\n/ngZ3pHj7Y4k3UiieR/N771GMtSM1+vlhhtu0PmkktW0yyGHlJeXx3/+538CEF6zjNZ9u21OJN1F\nbOt6Gv/1HMlQM/3792fWrFnU1NSoTCWrqVDlCx133HF86UtfAssk+OGbWMmk3ZEkg1mWScuK92l+\n/3WsZIJTTjmFmTNn6hql0iOoUOWwrrzySioqKkg27yO0eqndcSRDmbEoTe/8g0igFofDwbRp0/jm\nN79JTk6O3dFEuoQKVQ4rJyeHG2+8EcMwiNTV0rq3we5IkmFaG3ez719/p7VhGwUFBfzgBz/g3HPP\n1RCv9CgqVDkifr+f888/HyyL4JI3sRIJuyNJhohurqPxjfmY4RYGDx7MT37yE0aOHGl3LJEup0KV\nI3bppZf227doAAARjUlEQVRSXV1NsqWJ0KoldscRm1mmSctH7xFc8iaYSaZOncodd9xBaanOWZae\nSYUqR8zj8XDTTTfhcDiIrF1BfNd2uyOJTcxomKa3XyKybiVOp5Ovf/3r3HDDDboyjPRoKlTpkEGD\nBnHRRRcBEPxwIWZr3OZE0tVa9zaw7/W/07p7B8XFxdxxxx2ceeaZmi+VHk+FKh128cUXM2DAAMxw\nkNCK9+2OI10osmENjQtfwIyG8fv9zJ49m2HDhtkdSyQjqFClw1wuFzfddBNOp5PohjXEd261O5J0\nMiuZJLj0bVr+/TaYJmeffTa33Xabrlkqsh8VqhyVfv36cemllwIQXPoWZjxmcyLpLMlIiMa3XiS6\ncQ0ul4sbb7yR6667DpdLK5eK7E+FKkftggsuYMiQIZiREC21i+yOI50gvnsHja//ncTeBnr16sXM\nmTOZMmWK3bFEMpIKVY6a0+nkpptuwu12E9tcR2zbJrsjSZpYlkVk3Uqa3noJMxbhuOOOY9asWQwa\nNMjuaCIZS4Uqx6SqqoorrrgCgJZ/v40Zi9qcSI6VlUwQ/HAhLR+9B5bJ+eefzy233EJhYaHd0UQy\nmgpVjtk555zD8OHDMWMRWj561+44cgyS4SCNb84ntrkOj8fDzTffzNVXX43T6bQ7mkjGU6HKMXM4\nHNx4443k5OQQ27qe6Nb1dkeSoxBvqGff638n0biH8vJyfvjDH3LyySfbHUuk21ChSlqUl5dz9dVX\nA9Cy7B3MaNjmRHKkLMsiHKil6Z1/YMVjjB07llmzZtG/f3+7o4l0KypUSZszzzyT0aNHY8VjBP/9\nNpZl2R1JDsNKJAh+8K/UAh2WxUUXXcT3vvc9vF6v3dFEuh0VqqSNYRh84xvfIC8vj/j2zcQ2r7U7\nknwBMx6l8e2XiG1dT25uLt/+9re5/PLLcTj0siByNPSXI2nVq1cvpk2bBkBL7XskwyGbE8nBJCMh\nGhe+SGJvA6WlpfzoRz9i4sSJdscS6dZUqJJ2U6ZMYdy4cVitcYJLF2roN8MkW5pofPMFks376NOn\nDzNnzqS6utruWCLdngpV0s4wDKZPn47X66W1oZ7oxo/tjiRtEo17aHzzBcxwkMGDB3PnnXfSq1cv\nu2OJZAUVqnSK4uJirr/+egBCyxeTDAXtDSSpZQQXvoAZizBq1Chuu+02CgoK7I4lkjVUqNJpTjrp\nJCZNmoSVaCX44Zsa+rVRbPtmmt5+GSvRSk1NDd/73vfIzc21O5ZIVlGhSqe6/vrrKSwspHX3DiLr\nVtodp0eKbqqjedFrYCY588wz+da3voXb7bY7lkjWUaFKpyooKGD69OkAhFd+QCLYZHOiniW8dgXB\nD99sP8f0+uuv12kxIp1Ef1nS6SZMmMDkyZNTF6n+8E0sy7Q7UtazLIvQyiWE2i6rd80113D55Zdj\nGIbNyUSylwpVusS0adMoKSkhsbeBSN1yu+NkNcsyaVn2DuGPl7Wvs/zlL3/Z7lgiWU+FKl3C6/Xy\njW98A4DQqg9JBBttTpSdLDNJ8P1/Ed2wBpfLxbe//W1dEFyki6hQpcuMHTuWqVOngmnSorV+085K\ntNL07qvE6jeQl5fHLbfcwvjx4+2OJdJjqFClS1111VXtR/1GNwXsjpM1zFiUxrdeorWhnsLCQm6/\n/XZGjhxpdyyRHkWFKl3K5/O1r/UbWv4+ZjRic6LuLxkO0bjwBRL7dlFWVsaMGTMYOHCg3bFEehwV\nqnS5k046iTFjxmC1xmhZvsjuON1aIthE45vzSQYbqa6uZsaMGVRWVtodS6RHUqFKlzMMg+uvvx6P\nx0NsyzriO7bYHalbat23m8aF8zEjLQwdOpQ777yT0tJSu2OJ9FgqVLFFeXk5l156KQDBZe9gJVpt\nTtS9xHdto+mtF7FiUcaMGcOtt96Kz+ezO5ZIj6ZCFducd955DBgwADPcQmj1UrvjdBuxbRtpeucV\nrEQrJ554It/97ne1Lq9IBlChim2cTifTp0/HMAwia1eQaNxjd6SMF90YoHnxP8FMctZZZ3HzzTfj\ncrnsjiUiqFDFZoMHD+acc84ByyK49C0tS/gFwoFagksXgmVxySWXcN1112ldXpEMor9Gsd1ll11G\naWkpicbdRNatsjtOxrEsi5YV7xNa8T6QWsbxsssu07q8IhlGhSq2y8vL47rrrgMgvHIJyXCLzYky\nh2WlVpWKBGpxOBx885vf5Nxzz7U7logchApVMsL48eOpqanBSiZoWfauliUErGSS5sWvE934MW63\nm+985zuceuqpdscSkUNQoUrG+NrXvkZeXh7xHZuJb9todxxbma1xmt59hfi2jeTl5XPrrbdywgkn\n2B1LRL6AClUyRklJCVdeeSUALR+9hxmP2ZzIHmY0QtNbL9G6axtFRUXcccftDB8+3O5YInIYKlTJ\nKGeccQbDhg3DjIYJrfzA7jhdLhkK0rhwPonG3ZSXlzNz5kwGDBhgdywROQIqVMkoDoeD6dOn43Q6\niW5YQ+ueHXZH6jKJpr2pdXlbmhkwYAAzZsygvLzc7lgicoRUqJJx+vbtywUXXABAcOnbWGbS5kSd\nr3XPDhoXvoAZDTNixAhuv/12iouL7Y4lIh2gQpWMdPHFF1NRUUEy2Eg4UGt3nE4V276ZprdfxmqN\nM2HCBP7nf/6H/Px8u2OJSAepUCUjeTwevv71rwMQXrOMRLDJ5kSdI7q5juZFr2Elk0ydOpX/+q//\nwuPx2B1LRI6CClUy1qhRo5gyZQqYSVqWvZ1156aG65YTXPImWBYXXnghN9xwA06n0+5YInKUVKiS\n0a6++moKCgpo3bWd2OY6u+OkRWopwQ8ILV8MwDXXXMOVV16ppQRFujkVqmS0goICrr32WgBali/G\njEVsTnRsLNOkZelbRAIf4XA4uOmmm/jyl79sdywRSQMVqmS8U045hdGjR2PFY7TULrY7zlGzkgma\nF/+T6KYAHo+H73znO0yePNnuWCKSJipUyXiGYXD99dfjdruJbVlLfOdWuyN1mNkap+mdfxDfvon8\nfC0lKJKNVKjSLVRUVHDppZcCEFz2DlYiYXOiI2dGwzQtfJHW3TsoLinhzjvvxO/32x1LRNJMhSrd\nxnnnnUe/fv0wQ0FCa5baHeeIJEPN7HvzBRJNe6ioqGDmjBn069fP7lgi0glUqNJtuFwupk+fjmEY\nROqWk2jaY3ekL5Ro2kPjG/MxQ80MHDiQGTNm0Lt3b7tjiUgnUaFKtzJ06FDOPvtssKzUsoSWaXek\ng4rv3kHjwhcxYxGOO+44br/9doqKiuyOJSKdSIUq3c7ll19OSUkJiX27iK5fbXecz4lt29S+lGBN\nTQ3f//73ycvLszuWiHQyFap0O3l5efzHf/wHAKGVS0iGQzYn+lR0U4DmxQvATHLmmWfyrW99C7fb\nbXcsEekCKlTpliZOnMjEiROxEq20fPSu3XEACAdqCX64ECyLSy65hOuvvx6HQ39iIj2F/tql2/ra\n175Gbm4u8e2biNVvtC2HZVm0LF9MaMX77bkuu+wyLSUo0sOoUKXbKi0t5YorrgCg5aN3MVvjXZ7B\nMk2CHy4kUrcch9PJzTffzDnnnNPlOUTEfipU6dbOOusshgwZghkNE1q5pEu/t5VM0LxoAbHNdXg8\nHr733e9y8sknd2kGEckcKlTp1hwOB9OnT8fhdBJdv4rWPTs7/XtaZpJE8z4a336Z+I7NeL1ebrvt\nNsaOHdvp31tEMleHJnmsbLsgpWSNP//5z8yfPx9nYQklZ34V4xgPBrIsCzMSJtnSRLKlkWRLM8lg\nU+rrcBDa/hRKS0u55ZZbqK6uTsePISIZxujAwRAqVMkK8Xic2267jYaGBryjJpI//MgWnjfjsbbS\nbGovzERLE8mWZkgefL1gwzDo3bs3AwYM4JprrqGsrCydP4qIZBAVqvRIK1as4J577gGHk9KzL8Xp\nS61MZCUTJEPBT/cw20uzCSsWPeS/V1hYSGVlJZWVlVRVVbXflpeX69xSkR5ChSo91gMPPMA777yD\ns7AER66XZEsTZjh4yMd7PJ4DyvKTAq2srMTr9XZhchHJRCpU6bGam5u59dZbaWlpad/mcDjo3bv3\n5wqzqqqKkpISnS8qIoekQpUebePGjaxevZqKigqqqqro3bs3LpfL7lgi0g2pUEVERNKgI4Wq81BF\nRETSQIUqIiKSBipUERGRNFChioiIpIEKVUREJA1UqCIiImmgQhUREUkDFaqIiEgaqFBFRETSQIUq\nIiKSBipUERGRNFChioiIpIEKVUREJA1UqCIiImmgQhUREUkDFaqIiEgaqFBFRETSQIUqIiKSBipU\nERGRNFChioiIpIEKVUREJA1UqCIiImmgQhURERERERERERERERERERERERERERERERERERERERER\nERERERERERERERH5YsaRPMjv938H+C+gD7AemBUIBJ7qzGA9jd/vdwI/Au4AfhwIBH5sb6Ls4/f7\nPcBtwDVANbAR+F0gEPi9nbmyid/vLwZmAZcAFcAW4GHgrkAgYNmZLRv5/f5CYDUQDwQCg+zOk038\nfv9GoP9B7rovEAj898Ge4zqCf/Rm4C7gJuA94Hxgnt/v3xsIBF49+rjyCb/fXwH8idQLUALQC0/n\n+DVwJXAjsBT4CnCf3++PBgKBh21Nlj2eJvUidB2wAbgA+C0QAX5pY65s9ROgDKi3O0gWsoBftH3s\nL3yoJ3xhofr9fgO4Hfh9IBB4rG3zb/x+/9S27SrU9Lga2ANcBOyyOUtW8vv9RcA3gO8HAoG/tm3+\nrd/vPx+YRmovSo6B3+/vB9QAVwcCgX+1bf6d3++/CPj/UKGmld/vnwhMB54ETrc3TdZqCQQCDUf6\n4MPtoY4gNcz72eJcQOrFKCcQCMQ6GFA+75lAIPBrAL/fb3eWrBQIBJr8fn8Vn3932QCMtSFS1gkE\nAluA0oPcZQCtXRwnq7VNET0I/C9HOHUnne9whTq07XbjZ7avJ3Ut1cGkxu/lGAQCAQ3XdIFAILBn\n/6/9fn8+cCbwoj2Jspvf73eT2vufTGqoXdLnW4AXuJvUcReSAQ5XqAVtt6HPbG9puy1MbxyRLvU7\nUr/DP7M7SLbx+/3vApOA3cBVgUBgvs2Rsobf768mdeDXVwOBQKtGtTqNAdT4/f5XSI1ihYDHgbsD\ngUD8YE847EFJItmm7diA+4FrgcsDgcAGmyNloyuAXqSO9n3a7/ffEAgEnrQ5U7b4LfB8IBB43e4g\nWa4ByAPuAbaTmqf+GTAQ+PrBnnC4Qm1qu/3snmjRZ+4X6Rba5p4eBi4DLtOeU+cIBAJbga3AR36/\n3wf8/6QOnpFj4Pf7LwROA0bZnSXbBQKBSZ/ZtKLtNKWf+P3+OwKBwLbPPudwhVrXdjsEWLnf9mGk\nDjJYd7RhRWxyH6mjqc8NBAJv2x0mm/j9/gHAGcC8QCCQ2O+ulUCJ3+8v78gRk3JQl5I68Kt+v6Fe\nB2D4/f5WUuew/8SucD3AR223VcDnCtXxRc8MBAIBUgcgffkzd10ALAgEAjpyT7oNv99/I6mhmotU\npp1iGDAXmPKZ7WNIHV29t8sTZZ87Sf1/Hr/fxwOkXtyPJ3Xkrxwjv98/3O/3P+L3+wd/5q7xQJJU\nL37Okcyh/hj4Y9tBBguBq0iNJX/2j0aOkt/vPwEobvvSAQzy+/2nt339nk5NOnZtw44/A+YAH/v9\n/sr97w8EAjtsCZZdXgeWAA/5/f5vAWtJvVZ8E5jzmb1WOQptw4wH7Bn5/f5dQGsgEFhlT6qstBmY\nCjzl9/u/T2r6YirwP8AfAoHAvoM96Qv3UAECgcDjwP8ltSzex6SWbbskEAgsSk9uAX5F6sXodVJv\ncq5r+/yfpFZPkmM3gdSblv9D6gCDbft96LSlNAgEAiZwIak33o8Cy4HvkHrt+I59ybKehVZXS6tA\nIBAhNX2xDniK1Omht5M67/dbNkYTERERERERERERERERERERERERERERERERERERERERERERERE5\nSv8PDq8C2uF9io4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fdb90704490>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(8, 4.5))\n",
    "plt.xlim((0,5))\n",
    "sns.violinplot(resolutions.values, vert=False, color = sns.color_palette()[5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# function to map to a 0 to 1 scale\n",
    "# http://stackoverflow.com/questions/929103/convert-a-number-range-to-another-range-maintaining-ratio\n",
    "\n",
    "def remap( x, oMin, oMax, nMin, nMax ):\n",
    "\n",
    "    #range check\n",
    "    if oMin == oMax:\n",
    "        print \"Warning: Zero input range\"\n",
    "        return None\n",
    "\n",
    "    if nMin == nMax:\n",
    "        print \"Warning: Zero output range\"\n",
    "        return None\n",
    "\n",
    "    #check reversed input range\n",
    "    reverseInput = False\n",
    "    oldMin = min( oMin, oMax )\n",
    "    oldMax = max( oMin, oMax )\n",
    "    if not oldMin == oMin:\n",
    "        reverseInput = True\n",
    "\n",
    "    #check reversed output range\n",
    "    reverseOutput = False   \n",
    "    newMin = min( nMin, nMax )\n",
    "    newMax = max( nMin, nMax )\n",
    "    if not newMin == nMin :\n",
    "        reverseOutput = True\n",
    "\n",
    "    portion = (x-oldMin)*(newMax-newMin)/(oldMax-oldMin)\n",
    "    if reverseInput:\n",
    "        portion = (oldMax-x)*(newMax-newMin)/(oldMax-oldMin)\n",
    "\n",
    "    result = portion + newMin\n",
    "    if reverseOutput:\n",
    "        result = newMax - portion\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    3708.000000\n",
      "mean        0.541225\n",
      "std         0.196841\n",
      "min        -4.600000\n",
      "10%         0.400000\n",
      "25%         0.480000\n",
      "50%         0.560000\n",
      "75%         0.620000\n",
      "90%         0.672000\n",
      "max         0.822000\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "resolutions_mod = pd.Series([remap(x, 0, 5, 1, 0) for x in resolutions])\n",
    "resolutions_mod_stats = resolutions_mod.describe(percentiles = [.1,.25,.5,.75,.9])\n",
    "resolutions_mod_simple_cutoff_10 = resolutions_mod_stats['10%']\n",
    "resolutions_mod_simple_cutoff_25 = resolutions_mod_stats['25%']\n",
    "print resolutions_mod_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for_scoring['p_resolution_scaled'] = remap(for_scoring['p_resolution'], 0, 5, 1, 0)\n",
    "for_scoring['ssb_rez_score'] = for_scoring['p_resolution_scaled']/resolutions_mod_simple_cutoff_10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging scores & ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for_scoring['ssb_raw_score'] = for_scoring['ssb_rez_score'] + for_scoring['ssb_si_score']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import ranking as rk\n",
    "\n",
    "def ranker(ranks, reverse_flag):\n",
    "    pdb_rank = {}\n",
    "    \n",
    "    pdbs = [x[0] for x in ranks]\n",
    "    prop = [x[1] for x in ranks]\n",
    "    rank = list(rk.Ranking(prop, reverse=reverse_flag))\n",
    "\n",
    "    for i,v in enumerate(rank):\n",
    "        pdb_rank[pdbs[i]] = v[0]\n",
    "        \n",
    "    return pdb_rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "for_scoring['ssb_above_cutoffs'] = False\n",
    "for_scoring['ssb_rank'] = np.nan\n",
    "\n",
    "counter = 1\n",
    "for uniprot_id in tqdm(for_scoring.index.get_level_values(0).unique()):\n",
    "    \n",
    "    pdb_ids = for_scoring.loc[uniprot_id].index.tolist()\n",
    "    \n",
    "    # if there are no PDBs associated with a UniProt ID, skip this UniProt ID\n",
    "    if np.nan in pdb_ids:\n",
    "        continue\n",
    "    \n",
    "    pdbs_above_cutoffs = []\n",
    "    for pdb_id in pdb_ids:\n",
    "\n",
    "        ssb_rez_score = for_scoring.loc[(uniprot_id,pdb_id)].ssb_rez_score\n",
    "        ssb_si_score = for_scoring.loc[(uniprot_id,pdb_id)].ssb_si_score\n",
    "        \n",
    "        ssb_raw_score = for_scoring.loc[(uniprot_id,pdb_id)].ssb_raw_score\n",
    "        \n",
    "        if ssb_rez_score >= 1 and ssb_si_score >= 1:\n",
    "            for_scoring.loc[(uniprot_id,pdb_id), 'ssb_above_cutoffs'] = True\n",
    "            pdbs_above_cutoffs.append((pdb_id,ssb_raw_score))\n",
    "            \n",
    "\n",
    "    score_ordered = sorted(pdbs_above_cutoffs,key=lambda x: x[1], reverse=True)\n",
    "    score_rank_ordered = ranker(score_ordered, reverse_flag=False)\n",
    "    score_rank_ordered_list = [x[0] for x in sorted(score_rank_ordered.iteritems(), key=operator.itemgetter(1))]\n",
    "    \n",
    "    for pdb_id, raw_score in pdbs_above_cutoffs:\n",
    "        for_scoring.loc[(uniprot_id,pdb_id), 'ssb_rank'] = score_rank_ordered_list.index(pdb_id) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "      <th>p_resolution_scaled</th>\n",
       "      <th>ssb_rez_score</th>\n",
       "      <th>ssb_raw_score</th>\n",
       "      <th>ssb_above_cutoffs</th>\n",
       "      <th>ssb_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">O32583</th>\n",
       "      <th>1F0Z</th>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>1.208995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1ZUD</th>\n",
       "      <td>66</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>1.208995</td>\n",
       "      <td>0.604</td>\n",
       "      <td>1.51</td>\n",
       "      <td>2.718995</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">P00350</th>\n",
       "      <th>2ZYA</th>\n",
       "      <td>468</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "      <td>1.201245</td>\n",
       "      <td>0.680</td>\n",
       "      <td>1.70</td>\n",
       "      <td>2.901245</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2ZYD</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.993590</td>\n",
       "      <td>465</td>\n",
       "      <td>465</td>\n",
       "      <td>1.201245</td>\n",
       "      <td>0.700</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2.951245</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3FWN</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.995726</td>\n",
       "      <td>466</td>\n",
       "      <td>466</td>\n",
       "      <td>1.203828</td>\n",
       "      <td>0.700</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2.953828</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     u_seq_len  p_resolution  ssb_p_percent_seq_ident  \\\n",
       "u_uniprot_acc u_pdb                                                     \n",
       "O32583        1F0Z          66           NaN                 1.000000   \n",
       "              1ZUD          66          1.98                 1.000000   \n",
       "P00350        2ZYA         468          1.60                 0.993590   \n",
       "              2ZYD         468          1.50                 0.993590   \n",
       "              3FWN         468          1.50                 0.995726   \n",
       "\n",
       "                     ssb_p_aln_coverage  ssb_p_aln_coverage_sim  ssb_si_score  \\\n",
       "u_uniprot_acc u_pdb                                                             \n",
       "O32583        1F0Z                   66                      66      1.208995   \n",
       "              1ZUD                   66                      66      1.208995   \n",
       "P00350        2ZYA                  465                     465      1.201245   \n",
       "              2ZYD                  465                     465      1.201245   \n",
       "              3FWN                  466                     466      1.203828   \n",
       "\n",
       "                     p_resolution_scaled  ssb_rez_score  ssb_raw_score  \\\n",
       "u_uniprot_acc u_pdb                                                      \n",
       "O32583        1F0Z                   NaN            NaN            NaN   \n",
       "              1ZUD                 0.604           1.51       2.718995   \n",
       "P00350        2ZYA                 0.680           1.70       2.901245   \n",
       "              2ZYD                 0.700           1.75       2.951245   \n",
       "              3FWN                 0.700           1.75       2.953828   \n",
       "\n",
       "                    ssb_above_cutoffs  ssb_rank  \n",
       "u_uniprot_acc u_pdb                              \n",
       "O32583        1F0Z              False       NaN  \n",
       "              1ZUD               True         1  \n",
       "P00350        2ZYA               True         3  \n",
       "              2ZYD               True         2  \n",
       "              3FWN               True         1  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_scoring.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "      <th>p_resolution_scaled</th>\n",
       "      <th>ssb_rez_score</th>\n",
       "      <th>ssb_raw_score</th>\n",
       "      <th>ssb_above_cutoffs</th>\n",
       "      <th>ssb_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O32583</th>\n",
       "      <th>1ZUD</th>\n",
       "      <td>66</td>\n",
       "      <td>1.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>1.208995</td>\n",
       "      <td>0.604</td>\n",
       "      <td>1.51</td>\n",
       "      <td>2.718995</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00350</th>\n",
       "      <th>3FWN</th>\n",
       "      <td>468</td>\n",
       "      <td>1.50</td>\n",
       "      <td>0.995726</td>\n",
       "      <td>466</td>\n",
       "      <td>466</td>\n",
       "      <td>1.203828</td>\n",
       "      <td>0.700</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2.953828</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00363</th>\n",
       "      <th>1KF6</th>\n",
       "      <td>602</td>\n",
       "      <td>2.70</td>\n",
       "      <td>0.958472</td>\n",
       "      <td>577</td>\n",
       "      <td>577</td>\n",
       "      <td>1.158787</td>\n",
       "      <td>0.460</td>\n",
       "      <td>1.15</td>\n",
       "      <td>2.308787</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00370</th>\n",
       "      <th>4BHT</th>\n",
       "      <td>447</td>\n",
       "      <td>2.50</td>\n",
       "      <td>0.988814</td>\n",
       "      <td>442</td>\n",
       "      <td>442</td>\n",
       "      <td>1.195471</td>\n",
       "      <td>0.500</td>\n",
       "      <td>1.25</td>\n",
       "      <td>2.445471</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00448</th>\n",
       "      <th>1IXB</th>\n",
       "      <td>206</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.990291</td>\n",
       "      <td>204</td>\n",
       "      <td>205</td>\n",
       "      <td>1.197257</td>\n",
       "      <td>0.820</td>\n",
       "      <td>2.05</td>\n",
       "      <td>3.247257</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     u_seq_len  p_resolution  ssb_p_percent_seq_ident  \\\n",
       "u_uniprot_acc u_pdb                                                     \n",
       "O32583        1ZUD          66          1.98                 1.000000   \n",
       "P00350        3FWN         468          1.50                 0.995726   \n",
       "P00363        1KF6         602          2.70                 0.958472   \n",
       "P00370        4BHT         447          2.50                 0.988814   \n",
       "P00448        1IXB         206          0.90                 0.990291   \n",
       "\n",
       "                     ssb_p_aln_coverage  ssb_p_aln_coverage_sim  ssb_si_score  \\\n",
       "u_uniprot_acc u_pdb                                                             \n",
       "O32583        1ZUD                   66                      66      1.208995   \n",
       "P00350        3FWN                  466                     466      1.203828   \n",
       "P00363        1KF6                  577                     577      1.158787   \n",
       "P00370        4BHT                  442                     442      1.195471   \n",
       "P00448        1IXB                  204                     205      1.197257   \n",
       "\n",
       "                     p_resolution_scaled  ssb_rez_score  ssb_raw_score  \\\n",
       "u_uniprot_acc u_pdb                                                      \n",
       "O32583        1ZUD                 0.604           1.51       2.718995   \n",
       "P00350        3FWN                 0.700           1.75       2.953828   \n",
       "P00363        1KF6                 0.460           1.15       2.308787   \n",
       "P00370        4BHT                 0.500           1.25       2.445471   \n",
       "P00448        1IXB                 0.820           2.05       3.247257   \n",
       "\n",
       "                    ssb_above_cutoffs  ssb_rank  \n",
       "u_uniprot_acc u_pdb                              \n",
       "O32583        1ZUD               True         1  \n",
       "P00350        3FWN               True         1  \n",
       "P00363        1KF6               True         1  \n",
       "P00370        4BHT               True         1  \n",
       "P00448        1IXB               True         1  "
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_pdbs = for_scoring[(for_scoring.ssb_rank == 1)]\n",
    "best_pdbs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "FINAL_RESET = for_scoring.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DF_GEMPRO = DF_02_RXN_GENE_UNIPROT_PDB_HOMOL.merge(FINAL_RESET, how='outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>m_reaction</th>\n",
       "      <th>m_subsystem</th>\n",
       "      <th>m_formula</th>\n",
       "      <th>m_metabolites</th>\n",
       "      <th>m_gene_reaction_rule</th>\n",
       "      <th>m_gene</th>\n",
       "      <th>u_uniprot_acc</th>\n",
       "      <th>u_reviewed</th>\n",
       "      <th>u_gene_name</th>\n",
       "      <th>u_ec_number</th>\n",
       "      <th>u_description</th>\n",
       "      <th>u_seq</th>\n",
       "      <th>u_seq_len</th>\n",
       "      <th>u_pfam</th>\n",
       "      <th>u_kegg_id</th>\n",
       "      <th>u_refseq</th>\n",
       "      <th>u_go</th>\n",
       "      <th>u_pdb_count</th>\n",
       "      <th>u_pdb</th>\n",
       "      <th>p_experiment</th>\n",
       "      <th>p_resolution</th>\n",
       "      <th>p_chemicals</th>\n",
       "      <th>p_chains</th>\n",
       "      <th>p_chain_uniprot_map</th>\n",
       "      <th>p_ec_numbers</th>\n",
       "      <th>p_deposition_date</th>\n",
       "      <th>p_doi</th>\n",
       "      <th>p_pmid</th>\n",
       "      <th>p_space_group</th>\n",
       "      <th>i_entry_name</th>\n",
       "      <th>i_length</th>\n",
       "      <th>i_native</th>\n",
       "      <th>i_tm_helix</th>\n",
       "      <th>i_tm_score</th>\n",
       "      <th>ssb_p_percent_seq_ident</th>\n",
       "      <th>ssb_p_aln_coverage</th>\n",
       "      <th>ssb_p_aln_coverage_sim</th>\n",
       "      <th>ssb_si_score</th>\n",
       "      <th>p_resolution_scaled</th>\n",
       "      <th>ssb_rez_score</th>\n",
       "      <th>ssb_raw_score</th>\n",
       "      <th>ssb_above_cutoffs</th>\n",
       "      <th>ssb_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12PPDRtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_R_e &lt;=&gt; 12ppd_DASH_R_p</td>\n",
       "      <td>[12ppd_DASH_R_e, 12ppd_DASH_R_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>330</td>\n",
       "      <td>330</td>\n",
       "      <td>1.136662</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>2.136662</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12PPDStex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>12ppd_DASH_S_e &lt;=&gt; 12ppd_DASH_S_p</td>\n",
       "      <td>[12ppd_DASH_S_e, 12ppd_DASH_S_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>330</td>\n",
       "      <td>330</td>\n",
       "      <td>1.136662</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>2.136662</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>23CAMPtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>23camp_e &lt;=&gt; 23camp_p</td>\n",
       "      <td>[23camp_p, 23camp_e]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>330</td>\n",
       "      <td>330</td>\n",
       "      <td>1.136662</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>2.136662</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>23CCMPtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>23ccmp_e &lt;=&gt; 23ccmp_p</td>\n",
       "      <td>[23ccmp_e, 23ccmp_p]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>330</td>\n",
       "      <td>330</td>\n",
       "      <td>1.136662</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>2.136662</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23CGMPtex</td>\n",
       "      <td>Transport, Outer Membrane Porin</td>\n",
       "      <td>23cgmp_e &lt;=&gt; 23cgmp_p</td>\n",
       "      <td>[23cgmp_p, 23cgmp_e]</td>\n",
       "      <td>(b0241 or b0929 or b1377 or b2215)</td>\n",
       "      <td>b0241</td>\n",
       "      <td>P02932</td>\n",
       "      <td>1</td>\n",
       "      <td>phoE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['Outer membrane pore protein E']</td>\n",
       "      <td>MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...</td>\n",
       "      <td>351</td>\n",
       "      <td>['PF00267']</td>\n",
       "      <td>['ecj:JW0231', 'eco:b0241']</td>\n",
       "      <td>['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...</td>\n",
       "      <td>['GO:0009279; C:cell outer membrane; IEA:UniPr...</td>\n",
       "      <td>1</td>\n",
       "      <td>1PHO</td>\n",
       "      <td>X-RAY DIFFRACTION</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>['A']</td>\n",
       "      <td>{'A': ['P02932']}</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15-JAN-93</td>\n",
       "      <td>10.1038/358727A0</td>\n",
       "      <td>1380671</td>\n",
       "      <td>P 3 2 1</td>\n",
       "      <td>PHOE_ECOLI</td>\n",
       "      <td>351</td>\n",
       "      <td>1phoA</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.940171</td>\n",
       "      <td>330</td>\n",
       "      <td>330</td>\n",
       "      <td>1.136662</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1</td>\n",
       "      <td>2.136662</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  m_reaction                      m_subsystem  \\\n",
       "0  12PPDRtex  Transport, Outer Membrane Porin   \n",
       "1  12PPDStex  Transport, Outer Membrane Porin   \n",
       "2  23CAMPtex  Transport, Outer Membrane Porin   \n",
       "3  23CCMPtex  Transport, Outer Membrane Porin   \n",
       "4  23CGMPtex  Transport, Outer Membrane Porin   \n",
       "\n",
       "                           m_formula                     m_metabolites  \\\n",
       "0  12ppd_DASH_R_e <=> 12ppd_DASH_R_p  [12ppd_DASH_R_e, 12ppd_DASH_R_p]   \n",
       "1  12ppd_DASH_S_e <=> 12ppd_DASH_S_p  [12ppd_DASH_S_e, 12ppd_DASH_S_p]   \n",
       "2              23camp_e <=> 23camp_p              [23camp_p, 23camp_e]   \n",
       "3              23ccmp_e <=> 23ccmp_p              [23ccmp_e, 23ccmp_p]   \n",
       "4              23cgmp_e <=> 23cgmp_p              [23cgmp_p, 23cgmp_e]   \n",
       "\n",
       "                 m_gene_reaction_rule m_gene u_uniprot_acc  u_reviewed  \\\n",
       "0  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "1  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "2  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "3  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "4  (b0241 or b0929 or b1377 or b2215)  b0241        P02932           1   \n",
       "\n",
       "  u_gene_name u_ec_number                      u_description  \\\n",
       "0        phoE         NaN  ['Outer membrane pore protein E']   \n",
       "1        phoE         NaN  ['Outer membrane pore protein E']   \n",
       "2        phoE         NaN  ['Outer membrane pore protein E']   \n",
       "3        phoE         NaN  ['Outer membrane pore protein E']   \n",
       "4        phoE         NaN  ['Outer membrane pore protein E']   \n",
       "\n",
       "                                               u_seq  u_seq_len       u_pfam  \\\n",
       "0  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "1  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "2  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "3  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "4  MKKSTLALVVMGIVASASVQAAEIYNKDGNKLDVYGKVKAMHYMSD...        351  ['PF00267']   \n",
       "\n",
       "                     u_kegg_id  \\\n",
       "0  ['ecj:JW0231', 'eco:b0241']   \n",
       "1  ['ecj:JW0231', 'eco:b0241']   \n",
       "2  ['ecj:JW0231', 'eco:b0241']   \n",
       "3  ['ecj:JW0231', 'eco:b0241']   \n",
       "4  ['ecj:JW0231', 'eco:b0241']   \n",
       "\n",
       "                                            u_refseq  \\\n",
       "0  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "1  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "2  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "3  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "4  ['NP_414776.1', 'NC_000913.3', 'WP_000749863.1...   \n",
       "\n",
       "                                                u_go  u_pdb_count u_pdb  \\\n",
       "0  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "1  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "2  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "3  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "4  ['GO:0009279; C:cell outer membrane; IEA:UniPr...            1  1PHO   \n",
       "\n",
       "        p_experiment  p_resolution p_chemicals p_chains p_chain_uniprot_map  \\\n",
       "0  X-RAY DIFFRACTION             3         NaN    ['A']   {'A': ['P02932']}   \n",
       "1  X-RAY DIFFRACTION             3         NaN    ['A']   {'A': ['P02932']}   \n",
       "2  X-RAY DIFFRACTION             3         NaN    ['A']   {'A': ['P02932']}   \n",
       "3  X-RAY DIFFRACTION             3         NaN    ['A']   {'A': ['P02932']}   \n",
       "4  X-RAY DIFFRACTION             3         NaN    ['A']   {'A': ['P02932']}   \n",
       "\n",
       "  p_ec_numbers p_deposition_date             p_doi   p_pmid p_space_group  \\\n",
       "0          NaN         15-JAN-93  10.1038/358727A0  1380671       P 3 2 1   \n",
       "1          NaN         15-JAN-93  10.1038/358727A0  1380671       P 3 2 1   \n",
       "2          NaN         15-JAN-93  10.1038/358727A0  1380671       P 3 2 1   \n",
       "3          NaN         15-JAN-93  10.1038/358727A0  1380671       P 3 2 1   \n",
       "4          NaN         15-JAN-93  10.1038/358727A0  1380671       P 3 2 1   \n",
       "\n",
       "  i_entry_name  i_length i_native  i_tm_helix  i_tm_score  \\\n",
       "0   PHOE_ECOLI       351    1phoA           7         NaN   \n",
       "1   PHOE_ECOLI       351    1phoA           7         NaN   \n",
       "2   PHOE_ECOLI       351    1phoA           7         NaN   \n",
       "3   PHOE_ECOLI       351    1phoA           7         NaN   \n",
       "4   PHOE_ECOLI       351    1phoA           7         NaN   \n",
       "\n",
       "   ssb_p_percent_seq_ident  ssb_p_aln_coverage  ssb_p_aln_coverage_sim  \\\n",
       "0                 0.940171                 330                     330   \n",
       "1                 0.940171                 330                     330   \n",
       "2                 0.940171                 330                     330   \n",
       "3                 0.940171                 330                     330   \n",
       "4                 0.940171                 330                     330   \n",
       "\n",
       "   ssb_si_score  p_resolution_scaled  ssb_rez_score  ssb_raw_score  \\\n",
       "0      1.136662                  0.4              1       2.136662   \n",
       "1      1.136662                  0.4              1       2.136662   \n",
       "2      1.136662                  0.4              1       2.136662   \n",
       "3      1.136662                  0.4              1       2.136662   \n",
       "4      1.136662                  0.4              1       2.136662   \n",
       "\n",
       "  ssb_above_cutoffs  ssb_rank  \n",
       "0              True         1  \n",
       "1              True         1  \n",
       "2              True         1  \n",
       "3              True         1  \n",
       "4              True         1  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GEMPRO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_GEMPRO.to_pickle(os.path.join(DATA_FRAMES, 'DF_03_GEMPRO.pckl'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving the final GEM-PRO dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_GEMPRO['ssb_best_file'] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "STRUCT_BEST_EXP_PERFECT = os.path.join(STRUCT_FILES, 'best_set/experimental_perfect/')\n",
    "STRUCT_BEST_EXP_MUTATED = os.path.join(STRUCT_FILES, 'best_set/experimental_mutated/') \n",
    "STRUCT_BEST_HOMOLOGY = os.path.join(STRUCT_FILES, 'best_set/homology_models/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for directory in [STRUCT_BEST_EXP_PERFECT, STRUCT_BEST_EXP_MUTATED, STRUCT_BEST_HOMOLOGY]:\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "os.chdir(STRUCT_BEST_EXP_PERFECT)\n",
    "for uniprot_id,pdb_id in tqdm(best_pdbs.index.tolist()):\n",
    "    pdb_id = pdb_id.lower()\n",
    "    pdb_file = os.path.join(STRUCT_EXP_FILES, '%s.pdb' % pdb_id)\n",
    "    \n",
    "    if not os.path.exists(os.path.join(STRUCT_BEST_EXP_PERFECT, '%s.pdb' % pdb_id)):\n",
    "        shutil.copy(pdb_file, STRUCT_BEST_EXP_PERFECT)\n",
    "        \n",
    "    indices = DF_GEMPRO[(DF_GEMPRO.u_uniprot_acc == uniprot_id)].index.tolist()\n",
    "    for idx in indices:\n",
    "        DF_GEMPRO.loc[idx, 'ssb_best_file'] = '%s.pdb' % pdb_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "errors = []\n",
    "tmp2 = DF_GEMPRO[pd.notnull(DF_GEMPRO.i_entry_name)]\n",
    "for uniprot_id in tqdm(DF_GEMPRO[pd.isnull(DF_GEMPRO.ssb_best_file)].u_uniprot_acc.unique().tolist()):\n",
    "    \n",
    "    try:\n",
    "        itasser_name = tmp2[tmp2.u_uniprot_acc == uniprot_id].i_entry_name.unique()[0]\n",
    "    except:\n",
    "        errors.append(uniprot_id)\n",
    "        continue\n",
    "        \n",
    "    itasser_file = '%s_model1.pdb' % itasser_name\n",
    "    itasser_path = os.path.join(STRUCT_HOMOLOGY_FILES, itasser_file)\n",
    "    \n",
    "    if not os.path.exists(os.path.join(STRUCT_BEST_HOMOLOGY, itasser_file)):\n",
    "        shutil.copy(itasser_path, STRUCT_BEST_HOMOLOGY)\n",
    "        \n",
    "    indices = DF_GEMPRO[(DF_GEMPRO.u_uniprot_acc == uniprot_id)].index.tolist()\n",
    "    for idx in indices:\n",
    "        DF_GEMPRO.loc[idx, 'ssb_best_file'] = itasser_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['P0AD88', 'P39400']"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# these do not have a homology model\n",
    "errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DF_GEMPRO.to_pickle(os.path.join(DATA_FRAMES, 'DF_03_GEMPRO.pckl'))"
   ]
  }
 ],
 "metadata": {
  "_draft": {
   "nbviewer_url": "https://gist.github.com/ec22b728aeee2431586d"
  },
  "gist": {
   "data": {
    "description": "[212]_GEM-PRO_tutorial",
    "public": true
   },
   "id": "ec22b728aeee2431586d"
  },
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  },
  "notify_time": "5",
  "toc": {
   "toc_cell": true,
   "toc_number_sections": true,
   "toc_section_display": "block",
   "toc_threshold": 4,
   "toc_window_display": true
  },
  "toc_position": {
   "left": "2.25px",
   "right": "20px",
   "top": "42px",
   "width": "170px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
